<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Graphical Interface for Alternative Splicing Quantification, Analysis and
    Visualisation • psichomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Graphical Interface for Alternative Splicing Quantification, Analysis and
    Visualisation">
<meta property="og:description" content="Interactive R package with an intuitive Shiny-based graphical 
    interface for alternative splicing quantification and integrative analyses 
    of alternative splicing and gene expression based on The Cancer Genome Atlas
    (TCGA), the Genotype-Tissue Expression project (GTEx), Sequence Read Archive
    (SRA) and user-provided data. The tool interactively performs survival, 
    dimensionality reduction and median- and variance-based differential 
    splicing and gene expression analyses that benefit from the incorporation of
    clinical and molecular sample-associated features (such as tumour stage or 
    survival). Interactive visual access to genomic mapping and functional 
    annotation of selected alternative splicing events is also included.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script>
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">psichomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.12.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/GUI_tutorial.html">Visual interface tutorial</a>
    </li>
    <li>
      <a href="articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    </li>
    <li>
      <a href="articles/custom_data.html">Loading SRA and user-provided RNA-seq data</a>
    </li>
    <li>
      <a href="articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="psichomics" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#psichomics" class="anchor"></a>psichomics</h1></div>
<!-- badges: start -->

<blockquote>
<p><strong>Original article:</strong></p>
<p>Nuno Saraiva-Agostinho and Nuno L. Barbosa-Morais (2019). <a href="https://doi.org/10.1093/nar/gky888">psichomics: graphical application for alternative splicing quantification and analysis</a>. <em>Nucleic Acids Research</em>. 47(2), e7.</p>
</blockquote>
<p>Interactive R package with an intuitive Shiny-based graphical interface for alternative splicing quantification and integrative analyses of alternative splicing and gene expression based on <a href="https://tcga-data.nci.nih.gov">The Cancer Genome Atlas (TCGA)</a>, the <a href="http://www.gtexportal.org">Genotype-Tissue Expression (GTEx) project</a>, <a href="https://www.ncbi.nlm.nih.gov/sra">Sequence Read Archive (SRA)</a> and user-provided data.</p>
<p><em>psichomics</em> interactively performs survival, dimensionality reduction and median- and variance-based differential splicing and gene expression analyses that benefit from the incorporation of clinical and molecular sample-associated features (such as tumour stage or survival). Interactive visual access to genomic mapping and functional annotation of selected alternative splicing events is also included.</p>
<p><img src="reference/figures/screenshot.png" alt="Differential splicing analysis in psichomics"></p>
<div id="table-of-contents" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-contents" class="anchor"></a>Table of Contents</h2>
<ul>
<li>
<a href="#install-and-start-running">Install and start running</a>
<ul>
<li><a href="#bioconductor-release">Bioconductor release</a></li>
<li><a href="#github-version">GitHub version</a></li>
</ul>
</li>
<li><a href="#tutorials">Tutorials</a></li>
<li>
<a href="#workflow">Workflow</a>
<ul>
<li>
<a href="#data-input">Data input</a>
<ul>
<li><a href="#splicing-quantification">Splicing quantification</a></li>
<li><a href="#gene-expression-processing">Gene expression processing</a></li>
</ul>
</li>
<li><a href="#data-grouping">Data grouping</a></li>
<li><a href="#data-analyses">Data analyses</a></li>
</ul>
</li>
<li><a href="#feedback-and-support">Feedback and support</a></li>
<li><a href="#references">References</a></li>
</ul>
</div>
<div id="install-and-start-running" class="section level2">
<h2 class="hasAnchor">
<a href="#install-and-start-running" class="anchor"></a>Install and start running</h2>
<p>To install the package from <a href="https://www.bioconductor.org">Bioconductor</a>, type the following in <a href="https://www.rstudio.com/products/rstudio">RStudio</a> or in an R console:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">BiocManager<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"psichomics"</span>)</a></code></pre></div>
</div>
<div id="tutorials" class="section level2">
<h2 class="hasAnchor">
<a href="#tutorials" class="anchor"></a>Tutorials</h2>
<p>The following case studies and tutorials are available and were based on our <a href="https://doi.org/10.1093/nar/gky888">original article</a>:</p>
<ul>
<li><a href="http://rpubs.com/nuno-agostinho/psichomics-tutorial-visual">Visual interface</a></li>
<li><a href="http://rpubs.com/nuno-agostinho/psichomics-cli-tutorial">Command-line interface</a></li>
<li><a href="http://rpubs.com/nuno-agostinho/psichomics-custom-data">Loading SRA and user-provided RNA-seq data</a></li>
<li><a href="http://rpubs.com/nuno-agostinho/preparing-AS-annotation">Preparing alternative splicing annotations</a></li>
</ul>
</div>
<div id="workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow" class="anchor"></a>Workflow</h2>
<div id="data-input" class="section level3">
<h3 class="hasAnchor">
<a href="#data-input" class="anchor"></a>Data input</h3>
<p>Automatic retrieval and loading of pre-processed data from the following sources:</p>
<ul>
<li>
<a href="https://tcga-data.nci.nih.gov">TCGA</a> data of given tumours, including subject- and sample-associated information, junction quantification and gene expression data</li>
<li>
<a href="http://www.gtexportal.org">GTEx</a> data of given tissues, including subject- and sample-associated information, junction quantification and gene expression data</li>
<li>
<a href="https://www.ncbi.nlm.nih.gov/sra">SRA</a> data from select SRA projects via the <a href="https://jhubiostatistics.shinyapps.io/recount/">recount2</a> package</li>
</ul>
<p>Other SRA and user-provided data can be manually aligned and loaded. Please read <a href="http://rpubs.com/nuno-agostinho/psichomics-custom-data">Loading SRA and user-provided RNA-seq data</a>.</p>
<div id="splicing-quantification" class="section level4">
<h4 class="hasAnchor">
<a href="#splicing-quantification" class="anchor"></a>Splicing quantification</h4>
<p>The quantification of each alternative splicing event is based on the proportion of junction reads that support the inclusion isoform, known as percent spliced-in or PSI <a href="http://www.nature.com/nature/journal/v456/n7221/full/nature07509.html">(Wang <em>et al.</em>, 2008)</a>.</p>
<p>An estimate of this value is obtained based on the the proportion of reads supporting the inclusion of an exon over the reads supporting both the inclusion and exclusion of that exon. To measure this estimate, we require:</p>
<ol>
<li>Alternative splicing annotation: human (hg19 and hg38 assemblies) annotation is provided and custom annotations can be used.</li>
<li>Quantification of RNA-Seq reads aligning to splice junctions (junction quantification), either user-provided or retrieved from <a href="https://tcga-data.nci.nih.gov">TCGA</a>, <a href="http://www.gtexportal.org">GTEx</a> and <a href="https://www.ncbi.nlm.nih.gov/sra">SRA</a>.</li>
</ol>
</div>
<div id="gene-expression-processing" class="section level4">
<h4 class="hasAnchor">
<a href="#gene-expression-processing" class="anchor"></a>Gene expression processing</h4>
<p>Gene expression can be normalised, filtered and log2-transformed in-app or provided by the user.</p>
</div>
</div>
<div id="data-grouping" class="section level3">
<h3 class="hasAnchor">
<a href="#data-grouping" class="anchor"></a>Data grouping</h3>
<p>Molecular and clinical sample-associated attributes allow to establish groups that can be explored in data analyses.</p>
<p>For instance, <a href="https://tcga-data.nci.nih.gov">TCGA</a> data can be analysed based on smoking history, gender and race, among other attributes. Groups can also be manipulated (e.g. merged, intersected, etc.), allowing for complex attribute combinations. Groups can also be saved and loaded between sessions.</p>
</div>
<div id="data-analyses" class="section level3">
<h3 class="hasAnchor">
<a href="#data-analyses" class="anchor"></a>Data Analyses</h3>
<ul>
<li><p><strong>Dimensionality reduction</strong> via principal and independent component analysis (PCA and ICA) on alternative splicing quantification and gene expression.</p></li>
<li><p><strong>Differential splicing and gene expression analysis</strong> based on variance and median parametric and non-parametric statistical tests.</p></li>
<li><p><strong>Correlation between gene expression and splicing quantification</strong>, useful to correlate the expression of a given event with the expression of RNA-binding proteins, for instance.</p></li>
<li><p><strong>Survival analysis</strong> via Kaplan-Meier curves and Cox models based on sample-associated features. Additionally, we can study the impact of a splicing event (based on its quantification) or a gene (based on its gene expression) on patient survivability.</p></li>
<li><p><strong>Gene, transcript and protein annotation</strong> including relevant research articles</p></li>
</ul>
</div>
</div>
<div id="feedback-and-support" class="section level2">
<h2 class="hasAnchor">
<a href="#feedback-and-support" class="anchor"></a>Feedback and support</h2>
<p>Please send any feedback and questions on psichomics to:</p>
<blockquote>
<p>Nuno Saraiva-Agostinho (<a href="mailto:nunoagostinho@medicina.ulisboa.pt">nunoagostinho@medicina.ulisboa.pt</a>)</p>
<p><a href="http://imm.medicina.ulisboa.pt/group/distrans/">Disease Transcriptomics Lab, Instituto de Medicina Molecular (Portugal)</a></p>
</blockquote>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p>Wang, E. T., R. Sandberg, S. Luo, I. Khrebtukova, L. Zhang, C. Mayr, S. F. Kingsmore, G. P. Schroth, and C. B. Burge. 2008. <a href="http://www.nature.com/nature/journal/v456/n7221/full/nature07509.html"><em>Alternative isoform regulation in human tissue transcriptomes.</em></a> Nature 456 (7221): 470–76.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/psichomics">https://​www.bioconductor.org/​packages/​psichomics</a>
</li>
<li>Report a bug at <br><a href="https://github.com/nuno-agostinho/psichomics/issues">https://​github.com/​nuno-agostinho/​psichomics/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing psichomics</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Nuno Saraiva-Agostinho <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-5549-105X" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID"></a> </li>
<li>Nuno Luís Barbosa-Morais <br><small class="roles"> Author, lead, thesis advisor </small> <a href="https://orcid.org/0000-0002-1215-0538" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID"></a> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/nuno-agostinho/psichomics"><img src="https://travis-ci.org/nuno-agostinho/psichomics.svg?branch=master" alt="Travis Build Status"></a></li>
<li><a href="https://ci.appveyor.com/project/nuno-agostinho/psichomics"><img src="https://ci.appveyor.com/api/projects/status/github/nuno-agostinho/psichomics?branch=master&amp;svg=true" alt="AppVeyor Build Status"></a></li>
<li><a href="https://codecov.io/gh/nuno-agostinho/psichomics"><img src="https://codecov.io/gh/nuno-agostinho/psichomics/branch/master/graph/badge.svg" alt="codecov"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
