<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Interactive R package with an intuitive Shiny-based graphical 
    interface for alternative splicing quantification and integrative analyses 
    of alternative splicing and gene expression based on The Cancer Genome Atlas
    (TCGA), the Genotype-Tissue Expression project (GTEx), Sequence Read Archive
    (SRA) and user-provided data. The tool interactively performs survival, 
    dimensionality reduction and median- and variance-based differential 
    splicing and gene expression analyses that benefit from the incorporation of
    clinical and molecular sample-associated features (such as tumour stage or 
    survival). Interactive visual access to genomic mapping and functional 
    annotation of selected alternative splicing events is also included.">
<title>Visual interface for alternative splicing analysis and visualisation • psichomics</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Visual interface for alternative splicing analysis and visualisation">
<meta property="og:description" content="Interactive R package with an intuitive Shiny-based graphical 
    interface for alternative splicing quantification and integrative analyses 
    of alternative splicing and gene expression based on The Cancer Genome Atlas
    (TCGA), the Genotype-Tissue Expression project (GTEx), Sequence Read Archive
    (SRA) and user-provided data. The tool interactively performs survival, 
    dimensionality reduction and median- and variance-based differential 
    splicing and gene expression analyses that benefit from the incorporation of
    clinical and molecular sample-associated features (such as tumour stage or 
    survival). Interactive visual access to genomic mapping and functional 
    annotation of selected alternative splicing events is also included.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">psichomics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.24.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="articles/GUI_tutorial.html">Visual interface tutorial</a>
    <a class="dropdown-item" href="articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    <a class="dropdown-item" href="articles/custom_data.html">Loading user-provided data</a>
    <a class="dropdown-item" href="articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nuno-agostinho/psichomics/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="psichomics">psichomics<a class="anchor" aria-label="anchor" href="#psichomics"></a>
</h1></div>
<!-- badges: start -->

<blockquote>
<p><strong>Original article:</strong></p>
<p>Nuno Saraiva-Agostinho and Nuno L. Barbosa-Morais (2019). <a href="https://doi.org/10.1093/nar/gky888" class="external-link">psichomics: graphical application for alternative splicing quantification and analysis</a>. <em>Nucleic Acids Research</em>. 47(2), e7.</p>
</blockquote>
<p>Interactive R package with an intuitive Shiny-based graphical interface for alternative splicing quantification and integrative analyses of alternative splicing and gene expression based on <a href="https://tcga-data.nci.nih.gov" class="external-link">The Cancer Genome Atlas (TCGA)</a>, the <a href="http://www.gtexportal.org" class="external-link">Genotype-Tissue Expression (GTEx) project</a>, <a href="https://www.ncbi.nlm.nih.gov/sra" class="external-link">Sequence Read Archive (SRA)</a> and user-provided data.</p>
<p><em>psichomics</em> interactively performs: - Dimensionality reduction - Median- and variance-based differential splicing and gene expression analyses - Survival analysis - Correlation analysis - Grouping by clinical and molecular features (such as tumour stage or survival) - Genomic mapping and functional annotation of alternative splicing events and genes</p>
<div class="figure">
<img src="reference/figures/screenshot.png" alt=""><p class="caption">Differential splicing analysis in <em>psichomics</em></p>
</div>
<div class="section level2">
<h2 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h2>
<ul>
<li>
<a href="#install-and-start-running">Install and start running</a>
<ul>
<li><a href="#bioconductor">Bioconductor</a></li>
<li><a href="#github">GitHub</a></li>
<li><a href="#docker">Docker</a></li>
</ul>
</li>
<li><a href="#tutorials">Tutorials</a></li>
<li>
<a href="#workflow">Workflow</a>
<ul>
<li>
<a href="#data-input">Data input</a>
<ul>
<li><a href="#alternative-splicing-quantification">Alternative splicing quantification</a></li>
<li><a href="#gene-expression-processing">Gene expression processing</a></li>
</ul>
</li>
<li><a href="#data-grouping">Data grouping</a></li>
<li><a href="#data-analyses">Data analyses</a></li>
</ul>
</li>
<li><a href="#feedback-and-support">Feedback and support</a></li>
<li><a href="#references">References</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="install-and-start-running">Install and start running<a class="anchor" aria-label="anchor" href="#install-and-start-running"></a>
</h2>
<div class="section level3">
<h3 id="bioconductor">Bioconductor<a class="anchor" aria-label="anchor" href="#bioconductor"></a>
</h3>
<p>To install the package from <a href="https://www.bioconductor.org" class="external-link">Bioconductor</a>, type the following in <a href="https://www.rstudio.com/products/rstudio" class="external-link">RStudio</a> or in an R console:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"psichomics"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://nuno-agostinho.github.io/psichomics/">"psichomics"</a></span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>RStudio is now accessible via the web browser at <a href="https://localhost:8787" class="external-link uri">https://localhost:8787</a>
</li>
<li>Enter RStudio with user <code>rstudio</code> and password <code>bioc</code>
</li>
<li>Load psichomics using <code><a href="https://nuno-agostinho.github.io/psichomics/">library(psichomics)</a></code>
</li>
<li>Start the visual interface of psichomics with <code><a href="reference/psichomics.html">psichomics()</a></code>
</li>
</ol>
<p>Start the visual interface of psichomics with <code><a href="reference/psichomics.html">psichomics()</a></code></p>
</div>
<div class="section level3">
<h3 id="github">GitHub<a class="anchor" aria-label="anchor" href="#github"></a>
</h3>
<p>Install from GitHub (specify a branch or tag via the <code>ref</code> argument):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"nuno-agostinho/psichomics"</span>, ref<span class="op">=</span><span class="st">"master"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://nuno-agostinho.github.io/psichomics/">"psichomics"</a></span><span class="op">)</span></span></code></pre></div>
<p>Start the visual interface of psichomics with <code><a href="reference/psichomics.html">psichomics()</a></code></p>
</div>
<div class="section level3">
<h3 id="docker">Docker<a class="anchor" aria-label="anchor" href="#docker"></a>
</h3>
<p>The Docker images are based on <a href="https://www.bioconductor.org/help/docker/" class="external-link">Bioconductor Docker</a> and contain psichomics and its dependencies.</p>
<ol style="list-style-type: decimal">
<li>Pull the latest Docker image:</li>
</ol>
<pre><code>docker pull ghcr.io/nuno-agostinho/psichomics:latest</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Start RStudio Web from the Docker image:</li>
</ol>
<pre><code>docker run -e PASSWORD=bioc -p 8787:8787 ghcr.io/nuno-agostinho/psichomics:latest</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>Go to RStudio Web via the web browser at <a href="https://localhost:8787" class="external-link uri">https://localhost:8787</a>
</li>
<li>Log in RStudio with user <code>rstudio</code> and password <code>bioc</code>
</li>
<li>Load psichomics using <code><a href="https://nuno-agostinho.github.io/psichomics/">library(psichomics)</a></code>
</li>
<li>Start the visual interface of psichomics with <code><a href="reference/psichomics.html">psichomics()</a></code>
</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="tutorials">Tutorials<a class="anchor" aria-label="anchor" href="#tutorials"></a>
</h2>
<p>The following case studies and tutorials are available and were based on our <a href="https://doi.org/10.1093/nar/gky888" class="external-link">original article</a>:</p>
<ul>
<li><a href="https://nuno-agostinho.github.io/psichomics/articles/GUI_tutorial.html">Visual interface</a></li>
<li><a href="https://nuno-agostinho.github.io/psichomics/articles/CLI_tutorial.html">Command-line interface</a></li>
<li><a href="https://nuno-agostinho.github.io/psichomics/articles/custom_data.html">Loading user-provided data</a></li>
<li><a href="https://nuno-agostinho.github.io/psichomics/articles/AS_events_preparation.html">Preparing alternative splicing annotations</a></li>
</ul>
<p>Another tutorial was published as part of the Methods in Molecular Biology book series (the code for performing the analysis can be found <a href="https://github.com/nuno-agostinho/stem-cell-analysis-in-psichomics" class="external-link">here</a>):</p>
<blockquote>
<p>Nuno Saraiva-Agostinho and Nuno L. Barbosa-Morais (2020). <strong><a href="https://doi.org/10.1007/978-1-0716-0301-7_10" class="external-link">Interactive Alternative Splicing Analysis of Human Stem Cells Using psichomics</a></strong>. In: Kidder B. (eds) Stem Cell Transcriptional Networks. <em>Methods in Molecular Biology</em>, vol 2117. Humana, New York, NY</p>
</blockquote>
</div>
<div class="section level2">
<h2 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h2>
<div class="section level3">
<h3 id="data-input">Data input<a class="anchor" aria-label="anchor" href="#data-input"></a>
</h3>
<p>Automatic retrieval and loading of pre-processed data from the following sources:</p>
<ul>
<li>
<a href="https://tcga-data.nci.nih.gov" class="external-link">TCGA</a> data of given tumours, including subject- and sample-associated information, junction quantification and gene expression data</li>
<li>
<a href="http://www.gtexportal.org" class="external-link">GTEx</a> data of given tissues, including subject- and sample-associated information, junction quantification and gene expression data</li>
<li>
<a href="https://www.ncbi.nlm.nih.gov/sra" class="external-link">SRA</a> data from select SRA projects via the <a href="https://jhubiostatistics.shinyapps.io/recount/" class="external-link">recount</a> package</li>
</ul>
<p>Other SRA, <a href="https://github.com/vastgroup/vast-tools" class="external-link">VAST-TOOLS</a> and user-provided data can also be manually loaded. Please read <a href="https://nuno-agostinho.github.io/psichomics/articles/custom_data.html">Loading user-provided data</a> for more information.</p>
<div class="section level4">
<h4 id="alternative-splicing-quantification">Alternative splicing quantification<a class="anchor" aria-label="anchor" href="#alternative-splicing-quantification"></a>
</h4>
<p>The quantification of each alternative splicing event is based on the proportion of junction reads that support the inclusion isoform, known as percent spliced-in or PSI <a href="http://www.nature.com/nature/journal/v456/n7221/full/nature07509.html" class="external-link">(Wang <em>et al.</em>, 2008)</a>.</p>
<p>An estimate of this value is obtained based on the the proportion of reads supporting the inclusion of an exon over the reads supporting both the inclusion and exclusion of that exon. To measure this estimate, we require:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Alternative splicing annotation</strong>: human annotation is provided and custom annotations can be prepared for use in psichomics.</li>
<li>Quantification of RNA-Seq reads aligning to exon-exon splice junctions (<strong>exon-exon junction quantification</strong>), either user-provided or retrieved from <a href="https://tcga-data.nci.nih.gov" class="external-link">TCGA</a>, <a href="http://www.gtexportal.org" class="external-link">GTEx</a> and <a href="https://www.ncbi.nlm.nih.gov/sra" class="external-link">SRA</a>.</li>
</ol>
</div>
<div class="section level4">
<h4 id="gene-expression-processing">Gene expression processing<a class="anchor" aria-label="anchor" href="#gene-expression-processing"></a>
</h4>
<p>Gene expression can be normalised, filtered and log2-transformed in-app or provided by the user.</p>
</div>
</div>
<div class="section level3">
<h3 id="data-grouping">Data grouping<a class="anchor" aria-label="anchor" href="#data-grouping"></a>
</h3>
<p>Molecular and clinical sample-associated attributes allow to establish groups that can be explored in data analyses.</p>
<p>For instance, <a href="https://tcga-data.nci.nih.gov" class="external-link">TCGA</a> data can be analysed based on smoking history, gender and race, among other attributes. Groups can also be manipulated (e.g. merged, intersected, etc.), allowing for complex attribute combinations. Groups can also be saved and loaded between different sessions.</p>
</div>
<div class="section level3">
<h3 id="data-analyses">Data Analyses<a class="anchor" aria-label="anchor" href="#data-analyses"></a>
</h3>
<ul>
<li><p><strong>Dimensionality reduction</strong> via principal and independent component analysis (PCA and ICA) on alternative splicing quantification and gene expression.</p></li>
<li><p><strong>Differential splicing and gene expression analysis</strong> based on variance and median parametric and non-parametric statistical tests.</p></li>
<li><p><strong>Correlation between gene expression and splicing quantification</strong>, useful to correlate the expression of a given event with the expression of RNA-binding proteins, for instance.</p></li>
<li><p><strong>Survival analysis</strong> via Kaplan-Meier curves and Cox models based on sample-associated features. Additionally, we can study the impact of a splicing event (based on its quantification) or a gene (based on its expression) on patient survivability.</p></li>
<li><p><strong>Gene, transcript and protein annotation</strong>, including relevant research articles.</p></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="feedback-and-support">Feedback and support<a class="anchor" aria-label="anchor" href="#feedback-and-support"></a>
</h2>
<p>Please send any feedback and questions on psichomics to:</p>
<blockquote>
<p>Nuno Saraiva-Agostinho (<a href="mailto:nunoagostinho@medicina.ulisboa.pt">nunoagostinho@medicina.ulisboa.pt</a>)</p>
<p><a href="http://imm.medicina.ulisboa.pt/group/distrans/" class="external-link">Disease Transcriptomics Lab, Instituto de Medicina Molecular (Portugal)</a></p>
</blockquote>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Wang, E. T., R. Sandberg, S. Luo, I. Khrebtukova, L. Zhang, C. Mayr, S. F. Kingsmore, G. P. Schroth, and C. B. Burge. 2008. <a href="http://www.nature.com/nature/journal/v456/n7221/full/nature07509.html" class="external-link"><em>Alternative isoform regulation in human tissue transcriptomes.</em></a> Nature 456 (7221): 470–76.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/psichomics" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/nuno-agostinho/psichomics/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/nuno-agostinho/psichomics/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing psichomics</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Nuno Saraiva-Agostinho <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-5549-105X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Nuno Luís Barbosa-Morais <br><small class="roles"> Author, lead, thesis advisor </small> <a href="https://orcid.org/0000-0002-1215-0538" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/nuno-agostinho/psichomics/actions" class="external-link"><img src="https://github.com/nuno-agostinho/psichomics/workflows/R-CMD-check-bioc/badge.svg" alt="GitHub Actions Build status"></a></li>
<li><a href="https://codecov.io/gh/nuno-agostinho/psichomics" class="external-link"><img src="https://codecov.io/gh/nuno-agostinho/psichomics/branch/master/graph/badge.svg" alt="codecov"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
