<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 1.4 • psichomics</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Version 1.4" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">psichomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.14.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    </li>
    <li>
      <a href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    </li>
    <li>
      <a href="../articles/custom_data.html">Loading SRA, VAST-TOOLS and user-provided RNA-seq data</a>
    </li>
    <li>
      <a href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/nuno-agostinho/psichomics/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small>1.4</small></h1>
      
    </div>

    <div id="psichomics-140-22-oct-2017" class="section level1">
<h1 class="page-header" data-toc-text="1.4.0">
<a href="#psichomics-140-22-oct-2017" class="anchor"></a>psichomics 1.4.0 (22 Oct, 2017)</h1>
<ul>
<li>Support for loading new GTEx V7 data</li>
<li>Support gene expression data:
<ul>
<li>Load, filter, normalise and perform log2-transformation on gene expression data from TCGA</li>
<li>Perform principal component analysis based on gene expression data, survival analysis by gene expression cutoff and pairwise differential gene expression analysis</li>
<li>Correlate gene expression of a given gene against PSI values of multiple alternative splicing events</li>
</ul>
</li>
<li>Data loading:
<ul>
<li>Add step-wise instructions about loading of user-provided files</li>
<li>Filter GTEx junction quantification based on tissues of interest (all tissues are loaded by default)</li>
<li>Quantify splicing based on a list of genes (splicing events within all genes are quantified by default)</li>
<li>Parse sample information from TCGA samples using <code><a href="../reference/parseTcgaSampleInfo.html">parseTcgaSampleInfo()</a></code>
</li>
<li>Generate TCGA sample metadata when loading TCGA junction quantification</li>
<li>Present data summary after loading the data</li>
</ul>
</li>
<li>Data grouping:
<ul>
<li>Redesigned group creation and selection</li>
<li>Create groups based on genes and alternative splicing events</li>
<li>Assign a customisable colour per data group</li>
<li>Export or import patient and sample identifiers of data groups</li>
<li>Add new set operations when grouping (such as complement, subtraction and symmetric difference)</li>
<li>Suggest attributes of interest when creating groups</li>
<li>Allow to retrieve the universe of patient and sample identifiers by performing the complement group without any group selected</li>
<li>Statistically analyse group independence (useful to assess the overlap between a PCA cluster and groups derived from clinical and sample attributes, for instance)</li>
</ul>
</li>
<li>Differential analysis:
<ul>
<li>Label points based on top differentially spliced events or genes, selected alternative splicing events and/or selected genes</li>
<li>Create AS event and gene groups based on filtered or selected AS events and genes in the tables</li>
</ul>
</li>
<li>Dimensionality reduction techniques:
<ul>
<li>Subset data based on groups of AS events and genes before performing dimensionality reduction</li>
<li>Create data groups based on the partitioning clustering of PCA scores</li>
<li>Perform independent component analysis (ICA) on alternative splicing quantification and gene expression data</li>
</ul>
</li>
<li>Survival analysis:
<ul>
<li>Add p-value plot to visually infer the significance of survival analyses based on multiple alternative splicing quantification cutoffs</li>
</ul>
</li>
<li>Gene, transcript and protein information:
<ul>
<li>Information retrieval is now only dependent on a user-defined gene, instead of requiring alternative splicing quantification data to be loaded</li>
</ul>
</li>
</ul>
<div id="bug-fixes-and-other-improvements" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-and-other-improvements" class="anchor"></a>Bug fixes and other improvements</h2>
<ul>
<li>Show progress bar when running in the command-line interface</li>
<li>Fix inconsistent browser history navigation</li>
<li>Updated the CLI vignette with information on analysing gene expression data and a quick reference for functions</li>
<li>Update minimum version required of shiny (1.0.3)</li>
<li>Avoid replacing selected groups when manipulating new ones</li>
<li>Differential splicing analysis:
<ul>
<li>Fix data not being rendered in the table when zooming in the plot after data transformation was applied</li>
<li>Return p-value of NA instead of 0 when the value of Fligner-Killeen’s Test for Homogeneity of Variance is infinite</li>
<li>Discard value transformations that may return invalid data for the values chosen for the X and Y axes</li>
<li>Fix point that remains highlighted in the plot after deselecting the only selected row of the table</li>
<li>Improve readability of plot’s tooltip</li>
<li>Improve survival curves based on the optimal alternative splicing quantification cutoff:
<ul>
<li>Include the survival curve previews in 3 new columns within the differential splicing analyses table, instead of below that table; those columns consist of the survival curves, the optimal PSI cutoff and the respective log-rank test’s p-value</li>
<li>Allow to use survival data when plotting and table sorting</li>
<li>Include the optimal PSI cutoff and the respective log-rank test’s p-value in exported tables</li>
<li>Fix link to survival analyses using the previously calculated PSI cutoff</li>
</ul>
</li>
</ul>
</li>
<li>Principal component analysis:
<ul>
<li>When clicking on a alternative splicing event in the loadings plot, the appropriate differential splicing analyses will now be automatically rendered with the respective options, as expected</li>
</ul>
</li>
<li>Survival analysis:
<ul>
<li>Properly set the title of survival curves based on the selected splicing event’s quantification</li>
<li>Improve readability of Cox PH models</li>
<li>When performing survival analyses by alternative splicing cutoff, each patient is assigned the PSI value from the respective sample; for patients with more than one sample, the assigned sample is chosen based on the most frequent sample type across all patients (before, the first matched non-normal or non-control samples were used)</li>
</ul>
</li>
<li>Multiple other bug fixes and visual improvements</li>
</ul>
</div>
</div>
    <div id="psichomics-141-14-dec-2017" class="section level1">
<h1 class="page-header" data-toc-text="1.4.1">
<a href="#psichomics-141-14-dec-2017" class="anchor"></a>psichomics 1.4.1 (14 Dec, 2017)</h1>
<ul>
<li>GTEx data loading:
<ul>
<li>Add input elements to allow GTEx gene expression loading in the graphical interface</li>
<li>Fix bug that did not allow to select tissues to load GTEx v7 data (graphical interface)</li>
<li>Fix splicing events not being quantified based on GTEx v7 junction reads</li>
</ul>
</li>
<li>Gene expression normalisation:
<ul>
<li>Fix misleading gene expression (non-)normalisation by converting reads to counts per million (CPM) using <code><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">edgeR::cpm()</a></code> after normalisation using <code>edgeR::calcNormFactor()</code>
</li>
</ul>
</li>
<li>Alternative splicing quantification:
<ul>
<li>Updated support to properly parse new notation of alternative splicing annotation from Bioconductor (backwards compatible with older notation)</li>
<li>Raise error when no splicing events after quantification</li>
<li>Fix warning following the quantification of splicing events or its loading (incorrect parsing of gene information from splicing events)</li>
</ul>
</li>
<li>Dimensionality reduction:
<ul>
<li>Use the number (instead of the percentage) of tolerated missing values per sample as the argument to impute data from the remaining samples for those values before performing dimensionality reduction; by default, missing values are tolerated for 10 samples</li>
</ul>
</li>
<li>Update file description and README</li>
<li>Minor bug fixes and improvements</li>
</ul>
</div>
    <div id="psichomics-142-19-dec-2017" class="section level1">
<h1 class="page-header" data-toc-text="1.4.2">
<a href="#psichomics-142-19-dec-2017" class="anchor"></a>psichomics 1.4.2 (19 Dec, 2017)</h1>
<ul>
<li>Fix error when trying to load alternative splicing annotation (given updated hg19 and hg38 annotation that is now available for use with psichomics)</li>
</ul>
</div>
    <div id="psichomics-143-12-jan-2018" class="section level1">
<h1 class="page-header" data-toc-text="1.4.3">
<a href="#psichomics-143-12-jan-2018" class="anchor"></a>psichomics 1.4.3 (12 Jan, 2018)</h1>
<ul>
<li>Alternative splicing quantification:
<ul>
<li>Improve speed and memory usage when dealing with larger datasets</li>
<li>Improve quantification of alternative first and last exons: quantify alternative first and last exons based on all exon-exon junction reads that support each of the alternative exons</li>
<li>Print progress bar in R console</li>
</ul>
</li>
<li>Principal component analysis:
<ul>
<li>Change tolerated missing values per event to 5% by default</li>
<li>Show/save table with the contribution of events (for alternative splicing quantification) or genes (for gene expression) to the selected principal components</li>
<li>Add extra information when hovering variables in loading plot</li>
<li>Allow to plot top 100 variables that most contribute to the selected principal components (faster rendering of and interaction with loading plots)</li>
</ul>
</li>
<li>Differential analysis:
<ul>
<li>Allow to input a list of groups for the “group” argument of the functions <code><a href="../reference/diffAnalyses.html">diffAnalyses()</a></code> and <code><a href="../reference/plotDistribution.html">plotDistribution()</a></code> (command-line interface)</li>
<li>Allow to input a non-numeric vector or a row of a matrix/data frame in the “data” argument of the function <code><a href="../reference/plotDistribution.html">plotDistribution()</a></code> (command-line interface)</li>
</ul>
</li>
<li>Correlation between gene expression and alternative splicing:
<ul>
<li>Perform correlation between gene expression of multiple genes and quantification of multiple alternative splicing events</li>
</ul>
</li>
</ul>
<div id="bug-fixes-1" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-1" class="anchor"></a>Bug fixes</h2>
<ul>
<li>Fix unnamed events when only one event for a event type is returned</li>
<li>Minor copy-editing and overall improvements</li>
</ul>
</div>
</div>
    <div id="psichomics-144-12-feb-2018" class="section level1">
<h1 class="page-header" data-toc-text="1.4.4">
<a href="#psichomics-144-12-feb-2018" class="anchor"></a>psichomics 1.4.4 (12 Feb, 2018)</h1>
<ul>
<li>Update CITATION file to show citation to article in bioRxiv: <a href="https://www.biorxiv.org/content/early/2018/02/07/261180" class="uri">https://www.biorxiv.org/content/early/2018/02/07/261180</a>
</li>
<li>Update vignettes to include a case study based on the aforementioned article</li>
<li>Gene expression pipeline:
<ul>
<li>Perform <code>limma-trend</code> by default</li>
</ul>
</li>
<li>Alternative splicing quantification:
<ul>
<li>Quantification of alternative first and last exons: following more thorough testing, the new exon-centred method was considered to be less relevant to exploratory analysis (specially when compared with other types of events); as such, both methods are now available for quantification</li>
</ul>
</li>
<li>Dimensionality reduction:
<ul>
<li>Change tolerated missing values per event to 5% by default for both PCA and ICA (in both visual and command-line interfaces)</li>
<li>PCA: In the table that shows the events that most contribute to the selected principal components, show the rank</li>
</ul>
</li>
<li>Groups:
<ul>
<li>Automatically set sample type groups (i.e. normal, primary solid tumour, metastatic, etc.) for TCGA samples (visual interface only)</li>
</ul>
</li>
<li>Differential analysis:
<ul>
<li>Use numeric fields instead of sliders to precisely filter data</li>
<li>Fix table for differential expression not being filtered based on highlighted genes</li>
<li>Filter splicing events or genes to use when performing exploratory differential analyses</li>
</ul>
</li>
<li>Survival analysis:
<ul>
<li>Allow to stratify patients based on optimal gene expression cutoff</li>
<li>Select samples to be used for survival analysis</li>
</ul>
</li>
</ul>
<div id="bug-fixes" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes" class="anchor"></a>Bug fixes</h2>
<ul>
<li>Fix problems related with <code>DT</code> versions &gt;= 0.3:
<ul>
<li>Groups displayed as having the same attributes as last created group</li>
<li>Table not being updated in differential analyses according to event filtering based on the volcano plot</li>
</ul>
</li>
</ul>
</div>
</div>
    <div id="psichomics-145-4-apr-2018" class="section level1">
<h1 class="page-header" data-toc-text="1.4.5">
<a href="#psichomics-145-4-apr-2018" class="anchor"></a>psichomics 1.4.5 (4 Apr, 2018)</h1>
<ul>
<li>Mention psichomics manuscript throughout psichomics</li>
<li>Copy-edit graphical user interface and respective tutorial</li>
<li>Fix warnings and errors in Bioconductor</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: 'fa0226bf1a3df7e1c8ab8f0e70450e0a',
    indexName: 'stem-cell-analysis-in-psichomics',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


