<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 1.12 • psichomics</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Version 1.12" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script>


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">psichomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.12.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    </li>
    <li>
      <a href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    </li>
    <li>
      <a href="../articles/custom_data.html">Loading SRA and user-provided RNA-seq data</a>
    </li>
    <li>
      <a href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small>1.12</small></h1>
      
    </div>

    <div id="psichomics-1-12-1-29-january-2020" class="section level1">
<h1 class="page-header">
<a href="#psichomics-1-12-1-29-january-2020" class="anchor"></a>psichomics 1.12.1 (29 January, 2020)</h1>
<ul>
<li>Alternative splicing events can now be represented via diagrams:
<ul>
<li>Redesign of alternative splicing event selection (graphical interface)</li>
<li>
<code><a href="../reference/plotSplicingEvent.html">plotSplicingEvent()</a></code> plots diagram representation of alternative splicing events</li>
<li>In the visual interface, alternative splicing event diagrams were added below distribution plots (to quickly illustrate higher and lower values of alternative splicing quantification) and in annotation page</li>
</ul>
</li>
<li>User-provided junction quantification loading:
<ul>
<li>Support junction coordinates from mitochondrial, Z and W chromosomes</li>
<li>Fix issues with files containing splice junctions within <code>random</code>, <code>alt</code> and <code>unknown</code> chromosomes by discarding those rows (a warning is raised)</li>
</ul>
</li>
<li>Alternative splicing annotation:
<ul>
<li>
<code><a href="../reference/listSplicingAnnotations.html">listSplicingAnnotations()</a></code> can now be filtered by species, assembly and data of available annotations</li>
</ul>
</li>
<li>Improve import/export of data groups from/to a file, including colour support</li>
<li>Copy-edit and improve all tutorials, welcome message and help tab</li>
<li>Include link to article in Methods in Molecular Biology: <a href="https://doi.org/10.1007/978-1-0716-0301-7_10" class="uri">https://doi.org/10.1007/978-1-0716-0301-7_10</a>
</li>
</ul>
<div id="more-bug-fixes-and-minor-changes" class="section level2">
<h2 class="hasAnchor">
<a href="#more-bug-fixes-and-minor-changes" class="anchor"></a>More bug fixes and minor changes</h2>
<ul>
<li>TCGA and SRA data loading:
<ul>
<li>
<code><a href="../reference/loadTCGAdata.html">loadFirebrowseData()</a></code> now returns expected data when asking for multiple datasets (such as in the case of performing a pan-cancer analysis) in both visual and command-line interface</li>
<li>SRA projects containing only one column of extra information in sample metadata are now correctly loaded instead of raising an error (<code><a href="../reference/loadSRAproject.html">loadSRAproject()</a></code>)</li>
</ul>
</li>
<li>Data loading and manipulation:
<ul>
<li>Copy-edit information on the format of user-provided files</li>
<li>Warn when discarding rows with duplicated rownames after loading user-provided files</li>
<li>
<code><a href="../reference/getGtexDataTypes.html">getGtexDataTypes()</a></code> is now exported, as expected</li>
<li>
<code><a href="../reference/parseSplicingEvent.html">parseSplicingEvent()</a></code> now returns the coordinates as numeric if <code>coords = TRUE</code> and <code>char = FALSE</code>
</li>
<li>Improve dialog when trying to load a local folder without any supported files available</li>
</ul>
</li>
<li>Alternative splicing annotation:
<ul>
<li>Confirmation dialog is not displayed any more when creating a folder (specially useful while running the visual interface)</li>
<li>Allow to select cache directory of <code>AnnotationHub</code> (command-line interface)</li>
<li>Fix <code><a href="../reference/prepareSRAmetadata.html">prepareGeneQuant()</a></code> discarding the argument <code>strandedness</code> if either <code>stranded</code> or <code>stranded (reverse)</code>
</li>
</ul>
</li>
<li>Data grouping:
<ul>
<li>Samples not associated with any subject are now kept when exporting groups to a file</li>
<li>Fix issues related with importing groups based on a file with groups of splicing events and/or genes</li>
<li>Inform user when groups are successfully loaded from a file and whether any group elements are discarded in the process</li>
<li>Show number of genes contained in pre-made gene lists</li>
<li>Discard unavailable genes when creating group of genes based on pre-made gene lists (unless these are automatically created at startup)</li>
</ul>
</li>
<li>Density plot (<code><a href="../reference/plotDistribution.html">plotDistribution()</a></code>):
<ul>
<li>Fix visual bug when plotting a group with only one sample (if one data point is available for a group, only the rug plot is drawn)</li>
<li>After hiding all plot series, rug plots of the different groups can be distinguished based on the Y axis (different arbitrary Y values are given to each rug plot series)</li>
<li>Rug plot labels now show data values if sample names are not provided</li>
<li>Rug plot labels can now be rotated (rotation is not enabled by default given <code>Highcharts</code> issues that may occur at different zoom levels and depending on proximity between different sample values)</li>
<li>Fix misguiding example in function documentation</li>
<li>When hovering the values in the rug plot, the colour of the tooltip is now the same used for the rug points as expected</li>
</ul>
</li>
<li>Survival analysis (p-value plot):
<ul>
<li>Fix alternative splicing quantification cutoff being selected based on the one whose difference has the highest (instead of the lowest) p-value</li>
<li>Fix plot line label presenting “p &lt; 0.05” independently of the threshold used for significance</li>
</ul>
</li>
<li>Gene, transcript and protein annotation:
<ul>
<li>Show available annotation information and query PubMed even if Ensembl is down</li>
<li>Avoid app crash when searching for PubMed articles too many times</li>
</ul>
</li>
<li>Improve warning/error alerts:
<ul>
<li>Fix alerts crashing the visual interface</li>
<li>Improve message formatting regarding HTML-containing alerts</li>
</ul>
</li>
<li>Improve unit tests and function documentation</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#psichomics-1-12-1-29-january-2020">1.12.1</a></li>
      </ul>
    </div>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


