<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Version 1.12 • psichomics</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Version 1.12"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">psichomics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.24.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    <a class="dropdown-item" href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    <a class="dropdown-item" href="../articles/custom_data.html">Loading user-provided data</a>
    <a class="dropdown-item" href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nuno-agostinho/psichomics/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Version 1.12</h1>
      
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.12.1" id="psichomics-1121-29-january-2020">psichomics 1.12.1 (29 January, 2020)<a class="anchor" aria-label="anchor" href="#psichomics-1121-29-january-2020"></a></h2>
<ul><li>Alternative splicing events can now be represented via diagrams:<ul><li>Redesign of alternative splicing event selection (graphical interface)</li>
<li>
<code><a href="../reference/plotSplicingEvent.html">plotSplicingEvent()</a></code> plots diagram representation of alternative splicing events</li>
<li>In the visual interface, alternative splicing event diagrams were added below distribution plots (to quickly illustrate higher and lower values of alternative splicing quantification) and in annotation page</li>
</ul></li>
<li>User-provided junction quantification loading:<ul><li>Support junction coordinates from mitochondrial, Z and W chromosomes</li>
<li>Fix issues with files containing splice junctions within <code>random</code>, <code>alt</code> and <code>unknown</code> chromosomes by discarding those rows (a warning is raised)</li>
</ul></li>
<li>Alternative splicing annotation:<ul><li>
<code><a href="../reference/listSplicingAnnotations.html">listSplicingAnnotations()</a></code> can now be filtered by species, assembly and data of available annotations</li></ul></li>
<li>Improve import/export of data groups from/to a file, including colour support</li>
<li>Copy-edit and improve all tutorials, welcome message and help tab</li>
<li>Include link to article in Methods in Molecular Biology: <a href="https://doi.org/10.1007/978-1-0716-0301-7_10" class="external-link uri">https://doi.org/10.1007/978-1-0716-0301-7_10</a>
</li>
</ul><div class="section level3">
<h3 id="more-bug-fixes-and-minor-changes-1-12-1">More bug fixes and minor changes<a class="anchor" aria-label="anchor" href="#more-bug-fixes-and-minor-changes-1-12-1"></a></h3>
<ul><li>TCGA and SRA data loading:<ul><li>
<code><a href="../reference/loadTCGAdata.html">loadFirebrowseData()</a></code> now returns expected data when asking for multiple datasets (such as in the case of performing a pan-cancer analysis) in both visual and command-line interface</li>
<li>SRA projects containing only one column of extra information in sample metadata are now correctly loaded instead of raising an error (<code><a href="../reference/loadSRAproject.html">loadSRAproject()</a></code>)</li>
</ul></li>
<li>Data loading and manipulation:<ul><li>Copy-edit information on the format of user-provided files</li>
<li>Warn when discarding rows with duplicated rownames after loading user-provided files</li>
<li>
<code><a href="../reference/getGtexDataTypes.html">getGtexDataTypes()</a></code> is now exported, as expected</li>
<li>
<code><a href="../reference/parseSplicingEvent.html">parseSplicingEvent()</a></code> now returns the coordinates as numeric if <code>coords = TRUE</code> and <code>char = FALSE</code>
</li>
<li>Improve dialog when trying to load a local folder without any supported files available</li>
</ul></li>
<li>Alternative splicing annotation:<ul><li>Confirmation dialog is not displayed any more when creating a folder (specially useful while running the visual interface)</li>
<li>Allow to select cache directory of <code>AnnotationHub</code> (command-line interface)</li>
<li>Fix <code><a href="../reference/prepareSRAmetadata.html">prepareGeneQuant()</a></code> discarding the argument <code>strandedness</code> if either <code>stranded</code> or <code>stranded (reverse)</code>
</li>
</ul></li>
<li>Data grouping:<ul><li>Samples not associated with any subject are now kept when exporting groups to a file</li>
<li>Fix issues related with importing groups based on a file with groups of splicing events and/or genes</li>
<li>Inform user when groups are successfully loaded from a file and whether any group elements are discarded in the process</li>
<li>Show number of genes contained in pre-made gene lists</li>
<li>Discard unavailable genes when creating group of genes based on pre-made gene lists (unless these are automatically created at startup)</li>
</ul></li>
<li>Density plot (<code><a href="../reference/plotDistribution.html">plotDistribution()</a></code>):<ul><li>Fix visual bug when plotting a group with only one sample (if one data point is available for a group, only the rug plot is drawn)</li>
<li>After hiding all plot series, rug plots of the different groups can be distinguished based on the Y axis (different arbitrary Y values are given to each rug plot series)</li>
<li>Rug plot labels now show data values if sample names are not provided</li>
<li>Rug plot labels can now be rotated (rotation is not enabled by default given <code>Highcharts</code> issues that may occur at different zoom levels and depending on proximity between different sample values)</li>
<li>Fix misguiding example in function documentation</li>
<li>When hovering the values in the rug plot, the colour of the tooltip is now the same used for the rug points as expected</li>
</ul></li>
<li>Survival analysis (p-value plot):<ul><li>Fix alternative splicing quantification cutoff being selected based on the one whose difference has the highest (instead of the lowest) p-value</li>
<li>Fix plot line label presenting “p &lt; 0.05” independently of the threshold used for significance</li>
</ul></li>
<li>Gene, transcript and protein annotation:<ul><li>Show available annotation information and query PubMed even if Ensembl is down</li>
<li>Avoid app crash when searching for PubMed articles too many times</li>
</ul></li>
<li>Improve warning/error alerts:<ul><li>Fix alerts crashing the visual interface</li>
<li>Improve message formatting regarding HTML-containing alerts</li>
</ul></li>
<li>Improve unit tests and function documentation</li>
</ul></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

