<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 1.8 • psichomics</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Version 1.8" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">psichomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    </li>
    <li>
      <a href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    </li>
    <li>
      <a href="../articles/custom_data.html">Loading user-provided data</a>
    </li>
    <li>
      <a href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/nuno-agostinho/psichomics/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small>1.8</small></h1>
      
    </div>

    <div id="psichomics-181-3-december-2018" class="section level1">
<h1 class="page-header" data-toc-text="1.8.1">
<a href="#psichomics-181-3-december-2018" class="anchor"></a>psichomics 1.8.1 (3 December, 2018)</h1>
<ul>
<li>Add tag <code>ImmunoOncology</code> to BiocViews</li>
</ul>
</div>
    <div id="psichomics-182-26-march-2019" class="section level1">
<h1 class="page-header" data-toc-text="1.8.2">
<a href="#psichomics-182-26-march-2019" class="anchor"></a>psichomics 1.8.2 (26 March, 2019)</h1>
<ul>
<li>Data loading:
<ul>
<li>GTEx data can now be automatically downloaded and loaded on-demand</li>
<li>By default, data table now only displays at most the first 100 columns for performance reasons</li>
</ul>
</li>
<li>Alternative splicing quantification:
<ul>
<li>Allow to discard samples before alternative splicing quantification</li>
<li>In alternative splicing quantification dataset summary, plot quantification based on median, variance and range per splicing event across samples to provide tools to filter quantification (<code><a href="../reference/plotRowStats.html">plotRowStats()</a></code>)</li>
</ul>
</li>
<li>Gene expression filtering and normalisation:
<ul>
<li>Allow to discard samples before filtering and normalisation</li>
<li>Filter low read counts using <code><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">edgeR::filterByExpr()</a></code>
</li>
<li>Allow to perform <code><a href="https://rdrr.io/pkg/limma/man/voom.html">limma::voom()</a></code> on gene expression data (without design matrix) and to apply its normalisation methods</li>
<li>In gene expression dataset summary, plot distribution of gene expression per sample, distribution of library sizes and gene-wise mean and variance of gene expression across samples to provide the user tools to assess gene expression normalisation (<code><a href="../reference/plotGeneExprPerSample.html">plotGeneExprPerSample()</a></code>, <code><a href="../reference/plotDistribution.html">plotDistribution()</a></code> and <code><a href="../reference/plotRowStats.html">plotRowStats()</a></code>, respectively)</li>
<li>Convert between different gene identifiers (the original identifier is kept in some conditions, read <code><a href="../reference/convertGeneIdentifiers.html">convertGeneIdentifiers()</a></code>); in the visual interface, when filtering and normalising gene expression, ENSEMBL identifiers are converted to gene symbols, by default</li>
</ul>
</li>
<li>Groups:
<ul>
<li>By default, load pre-made lists of genes when loading gene expression or loading/performing alternative splicing quantification</li>
<li>Added pre-made list of genes that encode for RNA-binding proteins (Sebestyen et al. 2016), useful to postulate about the regulatory role of those proteins based on gene expression and PSI correlation analyses</li>
</ul>
</li>
<li>Correlation analyses:
<ul>
<li>Allow to use groups of genes and alternative splicing events in correlation analyses</li>
<li>Plot specific combinations of gene and alternative splicing events (<code>[.GEandAScorrelation</code>())</li>
<li>Display progress when performing correlation analyses</li>
<li>Display correlation results in a table (<code><a href="https://rdrr.io/r/base/table.html">as.table()</a></code>)</li>
</ul>
</li>
<li>Survival:
<ul>
<li>Render p-value plot by cutoff in command-line interface (<code><a href="../reference/plotSurvivalPvaluesByCutoff.html">plotSurvivalPvaluesByCutoff()</a></code>)</li>
</ul>
</li>
<li>Gene, transcript and protein information:
<ul>
<li>Modify keywords used to search for PubMed articles</li>
</ul>
</li>
</ul>
<div id="bug-fixes-and-minor-changes-2" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-and-minor-changes-2" class="anchor"></a>Bug fixes and minor changes</h2>
<ul>
<li>Improve console logging of error and warning alerts</li>
<li>Fix crash when loading psichomics with test data that is not locally available (by automatically downloading said data if not found)</li>
<li>Allow to edit file path in file/directory browser elements</li>
<li>Documentation:
<ul>
<li>Export functions mentioned in the documentation</li>
<li>Hide documentation of internal functions from the PDF reference manual</li>
</ul>
</li>
<li>Loading SRA data:
<ul>
<li>Accept a vector of files as the first argument (easier to use with <code><a href="https://rdrr.io/r/base/list.files.html">list.files()</a></code>)</li>
<li>Ask to overwrite file if one exists with the same name as the output file</li>
</ul>
</li>
<li>Groups:
<ul>
<li>Automatically set dropdown width for group attribute selection</li>
<li>Minor improvements to the group creation interface</li>
<li>Fix error when creating groups containing only samples and no matching subjects</li>
<li>Fix warning when displaying group preview only based on subjects or samples</li>
</ul>
</li>
<li>Alternative splicing quantification:
<ul>
<li>Automatically set the human genome version after loading data from TCGA (hg19), GTEx (hg19) or recount2 (hg38)</li>
<li>Fix progress bar</li>
<li>Decrease loading time after quantifying alternative splicing</li>
<li>By default, quantify skipped exons, mutually exclusive exons, alternative 3’ and 5’ splice sites, and alternative first and last exons; the default option is now consistent across both the visual and command-line interfaces)</li>
</ul>
</li>
<li>Differential analyses:
<ul>
<li>Allow distribution plots to show the name of the samples when hovering or when a rug plot is rendered if <code>rugLabels = TRUE</code> (function <code><a href="../reference/plotDistribution.html">plotDistribution()</a></code>)</li>
<li>Fix distribution plots requiring all samples in a group when using function <code><a href="../reference/plotDistribution.html">plotDistribution()</a></code>
</li>
<li>Fix group colours and opacity for rug plot points within the distribution plot (occurred in the command-line version; function <code><a href="../reference/plotDistribution.html">plotDistribution()</a></code>)</li>
<li>Fix wrong information in the table of differential splicing results (only occurs when the first splicing event is one for which there is not enough information to calculate statistical tests)</li>
<li>Fix inconsistency when presenting median and variance differences between gene expression and alternative splicing quantification</li>
<li>Fix error when groups contain samples outside the data being analysed</li>
</ul>
</li>
<li>Gene, transcript and protein information:
<ul>
<li>Fix article title formatting (e.g. bold and italics)</li>
</ul>
</li>
<li>Update psichomics citation with journal publication date</li>
</ul>
</div>
</div>
    <div id="psichomics-183-21-april-2019" class="section level1">
<h1 class="page-header" data-toc-text="1.8.3">
<a href="#psichomics-183-21-april-2019" class="anchor"></a>psichomics 1.8.3 (21 April, 2019)</h1>
<p>Replace deprecated <code>R.utils::evalWithTimeout()</code> with <code><a href="https://rdrr.io/pkg/R.utils/man/withTimeout.html">R.utils::withTimeout()</a></code></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: 'fa0226bf1a3df7e1c8ab8f0e70450e0a',
    indexName: 'stem-cell-analysis-in-psichomics',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


