<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 1.14 • psichomics</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Version 1.14" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">psichomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.18.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    </li>
    <li>
      <a href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    </li>
    <li>
      <a href="../articles/custom_data.html">Loading user-provided data</a>
    </li>
    <li>
      <a href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/nuno-agostinho/psichomics/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small>1.14</small></h1>
      
    </div>

    <div id="psichomics-1141-16-june-2020" class="section level1">
<h1 class="page-header" data-toc-text="1.14.1">
<a href="#psichomics-1141-16-june-2020" class="anchor"></a>psichomics 1.14.1 (16 June, 2020)</h1>
<p>Fix unit tests for R 4.0</p>
</div>
    <div id="psichomics-1142-11-august-2020" class="section level1">
<h1 class="page-header" data-toc-text="1.14.2">
<a href="#psichomics-1142-11-august-2020" class="anchor"></a>psichomics 1.14.2 (11 August, 2020)</h1>
<div id="support-for-loading-more-data-formats" class="section level2">
<h2 class="hasAnchor">
<a href="#support-for-loading-more-data-formats" class="anchor"></a>Support for loading more data formats</h2>
<ul>
<li>
<code><a href="../reference/getSplicingEventData.html">getSplicingEventData()</a></code>: get a table with all information for a given dataset of alternative splicing quantification data</li>
<li>VAST-TOOLS inclusion levels and gene expression tables:
<ul>
<li>Import VAST-TOOLS’ output files (inclusion levels, cRPKMs and gene read count tables); in the visual interface, use the <strong>User-provided data loading</strong> panel and load a specific alternative splicing quantification or load the folder containing such file; in the command-line interface, use <code><a href="../reference/loadLocalFiles.html">loadLocalFiles()</a></code>
</li>
<li>To parse splicing event from VAST-TOOLS in the command-line interface, use for instance <code><a href="../reference/parseSplicingEvent.html">parseSplicingEvent("HsaEX0007927", data=VASTTOOLSpsi)</a></code>
</li>
</ul>
</li>
<li>Support for importing inclusion tables with arbitrary alternative splicing event identifiers (information for these events will not be available, such as event type and cognate gene)</li>
<li>SRA metadata:
<ul>
<li>Automatically load metadata from the SRA Run Selector (usually comes in files named <code>SraRunTable.txt</code>) avoiding the need to use <code><a href="../reference/prepareSRAmetadata.html">prepareSRAmetadata()</a></code> first</li>
</ul>
</li>
<li>GTEx data loading (<code><a href="../reference/loadGtexData.html">loadGtexData()</a></code>):
<ul>
<li>Support for loading GTEx data from V8, V7, V6 or V4 releases</li>
<li>Organise GTEx data into folders named based on release version</li>
</ul>
</li>
<li>Copy-edit tutorial on loading user-provided data</li>
</ul>
</div>
<div id="new-features" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features" class="anchor"></a>New features</h2>
<ul>
<li>
<code><a href="../reference/plotSplicingEvent.html">plotSplicingEvent()</a></code>:
<ul>
<li>Alternative splicing diagrams now render automatically instead of showing SVG code (if printing more than one event, a table is displayed with event identifiers and respective diagrams)</li>
<li>Plot intron retention events (e.g. from VAST-TOOLS)</li>
</ul>
</li>
<li>
<code><a href="../reference/plotLibrarySize.html">plotLibrarySize()</a></code>: plot library size from gene expression data</li>
<li>Alternative splicing quantification filtering (visual interface):
<ul>
<li>New panel to allow filtering already loaded alternative splicing quantification based on event types, samples, PSI statistics (such as median, variance and range) and cognate genes (some filtering steps may not be available when using user-provided tables with inclusion levels)</li>
<li>Filter data based on individual PSI values (also available when quantifying alternative splicing)</li>
<li>Filter VAST-TOOLS events based on its quality scores for read coverage</li>
<li>Toggle specific statistics based on alternative splicing quantification values when filtering</li>
<li>Preview the original and filtered events in an easy-to-use plot</li>
</ul>
</li>
<li>Gene expression filtering:
<ul>
<li>Toggle specific statistics based on gene expression values when filtering (visual interface)</li>
<li>Arguments used in <code><a href="../reference/filterGeneExpr.html">filterGeneExpr()</a></code> and <code><a href="../reference/normaliseGeneExpression.html">normaliseGeneExpression()</a></code> are now returned as attributes of those functions</li>
</ul>
</li>
<li>Differential analysis:
<ul>
<li>Tooltips of volcano plots now include diagram of alternative splicing events</li>
</ul>
</li>
<li>Gene, transcript and protein annotation (visual interface):
<ul>
<li>Suggest genes based on loaded data</li>
<li>Change species and assembly when fetching information</li>
<li>Improve interface of PubMed article query, including persistent user-inputted tags even when changing selected gene</li>
<li>Improve relevance of results from PubMed articles</li>
</ul>
</li>
<li>Much faster calculation of row/column-wise means, medians, variances and ranges (helpful for plotting statistics of large datasets)</li>
<li>Include Dockerfile to create Docker images based on code revisions</li>
</ul>
</div>
<div id="bug-fixes-and-minor-changes-1" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-and-minor-changes-1" class="anchor"></a>Bug fixes and minor changes</h2>
<ul>
<li>Progress bar is now animated as in previous versions</li>
<li>Data loading (visual interface):
<ul>
<li>Allow to discard selected files in file browser input elements</li>
<li>When creating groups by sample index/identifiers, suggest sample names not only obtained from sample information, but also from alternative splicing quantification and gene expression datasets</li>
<li>“Browse…” button now opens file browser to select folder where data is stored (as expected) in GTEx and SRA panels</li>
<li>Only pre-create groups of genes (based on literature-based gene lists) if at least one of its genes exists in any of the loaded datasets</li>
<li>Correctly parse gene symbols containing underscores</li>
<li>Fix gene expression summary plots not showing in specific situations</li>
<li>Fix library size plot not working properly and causing rendering issues</li>
<li>Fix settings used to quantify alternative splicing not showing up</li>
<li>Show errors raised while reading a file (e.g. if file is too big for available memory)</li>
<li>Show alert if no GTEx data options are selected</li>
<li>Show filename of the file used to load gene expression and alternative splicing data from GTEx and SRA</li>
<li>Show helpful context messages in panel interfaces</li>
<li>Improve visual interface and minor copy-editing</li>
</ul>
</li>
<li>Local data loading:
<ul>
<li>Support loading data from GTEx V8 or previous releases</li>
<li>Fix bad formatting of help tooltips when using <code>shiny 1.4.0</code> or newer (visual interface)</li>
</ul>
</li>
<li>Gene expression filtering:
<ul>
<li>Hide message regarding the usage of no design matrix</li>
</ul>
</li>
<li>Alternative splicing annotation:
<ul>
<li>Try to load cached alternative splicing annotation if a timeout occurs</li>
<li>Include gene symbols in custom annotations if available</li>
<li>Account for possible filename changes when parsing annotations from VAST-TOOLS, rMATS, SUPPA and MISO</li>
</ul>
</li>
<li>Alternative splicing event selection (visual interface):
<ul>
<li>Include loading indicator while searching for events</li>
<li>Decrease number of operations performed after selecting an event</li>
<li>Fix crash when changing to an alternative splicing quantification dataset without the selected splicing event</li>
<li>Allow to search using an event identifier directly</li>
<li>Show event identifier instead of prettier identifier to avoid confusion</li>
</ul>
</li>
<li>Data grouping (visual interface):
<ul>
<li>Simplify group selection interface</li>
<li>Fix suggested attributes and index/identifiers in group creation not being cleared when changing to datasets where such data is unavailable (thus showing the attributes/index/identifiers of the previous dataset)</li>
<li>Show an alert when there is no data to create groups</li>
</ul>
</li>
<li>Dimensionality reduction:
<ul>
<li>
<code><a href="../reference/performPCA.html">performPCA()</a></code> and <code><a href="../reference/performICA.html">performICA()</a></code> now directly raise errors (instead of simply capturing and returning them)</li>
<li>PCA: loading plots now show parsed information of events (cognate gene, event type and genomic position) if available</li>
<li>PCA: the correct splicing event is now selected when clicking on any loadings in the loadings plot (visual interface)</li>
</ul>
</li>
<li>Differential analyses (visual interface):
<ul>
<li>Use group colours in density plots of differential expression table</li>
<li>Fix occasional crash when performing differential analyses with different number of groups during the same session</li>
<li>Tooltips of volcano plots are now properly positioned in high-resolution screens</li>
<li>The correct splicing event is now selected when clicking the density plots or survival curves in the table</li>
<li>When creating a group from differential splicing results, correctly set cognate genes of alternative splicing events</li>
<li>
<code><a href="../reference/diffAnalyses.html">diffAnalyses()</a></code>: deprecated <code>psi</code> argument was now removed</li>
</ul>
</li>
<li>Distribution plots (<code><a href="../reference/plotDistribution.html">plotDistribution()</a></code>):
<ul>
<li>After hiding all plot series, hide Y axis (rug plots of the different groups have arbitrary Y values to easily distinguish them)</li>
<li>Rug plots of gene expression density plots are now placed near the X axis as expected</li>
</ul>
</li>
<li>Correlation analyses (visual interface):
<ul>
<li>Warn when selecting genes that are not available in the selected gene expression dataset (instead of crashing the app)</li>
</ul>
</li>
<li>
<code><a href="https://rdrr.io/r/base/print.html">print()</a></code> extended to better display information on gene list objects; e.g. <code><a href="https://rdrr.io/r/base/print.html">print(getGeneList())</a></code>
</li>
<li>Fix issues when installing the package:
<ul>
<li>Fix error when unit testing in R 4.0 or higher (strings in data frames are not converted to factors by default)</li>
<li>Fix <code>R CMD check</code> warning of Unicode symbol translation in Windows</li>
<li>Fix comparing signed and unsigned integers in <code>Rcpp</code> functions</li>
</ul>
</li>
</ul>
</div>
</div>
    <div id="psichomics-1143-21-august-2020" class="section level1">
<h1 class="page-header" data-toc-text="1.14.3">
<a href="#psichomics-1143-21-august-2020" class="anchor"></a>psichomics 1.14.3 (21 August, 2020)</h1>
<ul>
<li>Fix unit tests due to changes in Ensembl API response</li>
</ul>
</div>
    <div id="psichomics-1144-30-august-2020" class="section level1">
<h1 class="page-header" data-toc-text="1.14.4">
<a href="#psichomics-1144-30-august-2020" class="anchor"></a>psichomics 1.14.4 (30 August, 2020)</h1>
<ul>
<li>Improve file browser dialog:
<ul>
<li>Now correctly allows to select a single file in macOS</li>
<li>Now correctly allows to select a directory in Windows</li>
<li>Correctly set default location when opening a file in Linux</li>
<li>Internal improvements</li>
</ul>
</li>
<li>Data loading:
<ul>
<li>Improve performance when loading compressed data (GZ and BZ2 files)</li>
<li>Increase optional verbosity while loading files</li>
<li>Speed up loading of VAST-TOOLS’ inclusion levels</li>
<li>Fix loading of VAST-TOOLS’ inclusion levels sometimes returning an error regarding duplicated row names</li>
<li>Fix issues with missing values as alternative splicing event complexity in VAST-TOOLS’ inclusion levels</li>
</ul>
</li>
<li>Copy-edit tutorials on loading user-provided data and using the command-line version</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: 'fa0226bf1a3df7e1c8ab8f0e70450e0a',
    indexName: 'stem-cell-analysis-in-psichomics',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


