<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Version 1.6 • psichomics</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Version 1.6"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">psichomics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.24.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    <a class="dropdown-item" href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    <a class="dropdown-item" href="../articles/custom_data.html">Loading user-provided data</a>
    <a class="dropdown-item" href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nuno-agostinho/psichomics/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Version 1.6</h1>
      
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.6.1" id="psichomics-161-5-july-2018">psichomics 1.6.1 (5 July, 2018)<a class="anchor" aria-label="anchor" href="#psichomics-161-5-july-2018"></a></h2>
<ul><li>Improve support for analysing SRA and user-provided data:<ul><li>New tutorial on performing alternative splicing analysis using SRA and user-provided data from FASTQ files (other tutorials were also updated)</li>
<li>Process gene and splice junction counts from STAR output for subsequent loading into psichomics</li>
<li>Automatically download and load data from select SRA projects using the recount R package</li>
</ul></li>
<li>Data grouping:<ul><li>Preview groups based on selected attribute before creation</li></ul></li>
<li>Gene, transcript and protein information:<ul><li>Display alternative splicing events in transcript plot</li></ul></li>
<li>Correlation between gene expression and alternative splicing:<ul><li>Allow to change plot height and font size</li></ul></li>
</ul><div class="section level3">
<h3 id="bug-fixes-and-minor-changes-1-6-1">Bug fixes and minor changes<a class="anchor" aria-label="anchor" href="#bug-fixes-and-minor-changes-1-6-1"></a></h3>
<ul><li>Support new version of Human (hg38 assembly) alternative splicing annotation (fixes wrong coordinates for many minus-strand splicing events)</li>
<li>Fix issues with TCGA sample metadata:<ul><li>No sample metadata loaded when loading TCGA files from local folder</li>
<li>Not retrieving sample metadata for all TCGA samples when multiple junction quantification files are loaded</li>
</ul></li>
<li>Data grouping:<ul><li>Correctly suggest how to select variables in subset expression and give more examples on how to use this feature</li>
<li>Fix subject/sample attributes not being updated for selection for the <code>GREP</code> interface</li>
<li>Fix error alerts not appearing when trying to create groups based on invalid parameters</li>
<li>Add an error alert when trying to extract group data from an user-provided file without such information</li>
</ul></li>
<li>Alternative splicing quantification:<ul><li>Fix crash when quantifying splicing for a single sample</li></ul></li>
<li>Differential splicing analysis (exploratory):<ul><li>Fix crash when inputting non-numeric values or minimum larger than the maximum limit as the axes for the splicing scatter plot</li></ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.6.2" id="psichomics-162-2-october-2018">psichomics 1.6.2 (2 October, 2018)<a class="anchor" aria-label="anchor" href="#psichomics-162-2-october-2018"></a></h2>
<ul><li>Update citations to link to article in Nucleic Acids Research: <a href="https://doi.org/10.1093/nar/gky888" class="external-link uri">https://doi.org/10.1093/nar/gky888</a>
</li>
<li>Copy-edit README, tutorials, DESCRIPTION, NEWS and code</li>
<li>Update screen shot</li>
<li>Add metadata regarding loaded datasets (for instance, path and name of files used for dataset loading, options set upon dataset creation, etc.)</li>
<li>Differential splicing and expression analysis:<ul><li>Allow to input formulas when highlighting points; this is an improvement when plotting -log10(q-values), for instance, where it is now possible to highlight values &gt;= -log10(0.05), i.e. q-values &lt;= 0.5, instead of inputting an approximate value</li></ul></li>
<li>Correlation analyses:<ul><li>Colour samples by group</li>
<li>Render contour based on a density estimate</li>
<li>Display exportable table with correlation analyses</li>
</ul></li>
<li>Groups:<ul><li>Show group colour in group selection element</li></ul></li>
<li>Retrieve articles from PubMed instead of PMC</li>
</ul><div class="section level3">
<h3 id="bug-fixes-and-minor-changes-1-6-2">Bug fixes and minor changes<a class="anchor" aria-label="anchor" href="#bug-fixes-and-minor-changes-1-6-2"></a></h3>
<ul><li>Improve file browser input in remote servers:<ul><li>Allow to directly input a path without using the file browser dialog</li>
<li>Display warning when calling the file browser dialog in RStudio Server or in an unsupported system</li>
</ul></li>
<li>TCGA data:<ul><li>Inform of sequencing technology used to obtain gene expression data</li></ul></li>
<li>GTEx data:<ul><li>Fix issues in downstream analyses when using GTEx data subset by tissues</li></ul></li>
<li>Differential analysis:<ul><li>Avoid limiting X axis of density plot when values outside 0 and 1 are provided, as in the case of gene expression (<code><a href="../reference/plotDistribution.html">plotDistribution()</a></code> function)</li>
<li>Fix filenames of exported tables from differential expression analysis incorrectly mentioning differential splicing analysis instead</li>
</ul></li>
<li>Correlation analysis:<ul><li>When plotting correlation analysis, loess curve was performed based on gaussian fitting, independent of the “family” argument of <code><a href="../reference/plot.GEandAScorrelation.html">plotCorrelation()</a></code>
</li></ul></li>
<li>Improve console logging of error and warning messages</li>
</ul></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

