<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Perform statistical analyses"><title>Perform statistical analyses — diffAnalyses • psichomics</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform statistical analyses — diffAnalyses"><meta property="og:description" content="Perform statistical analyses"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">psichomics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.24.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    <a class="dropdown-item" href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    <a class="dropdown-item" href="../articles/custom_data.html">Loading user-provided data</a>
    <a class="dropdown-item" href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nuno-agostinho/psichomics/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Perform statistical analyses</h1>
      <small class="dont-index">Source: <a href="https://github.com/nuno-agostinho/psichomics/blob/HEAD/R/analysis.R" class="external-link"><code>R/analysis.R</code></a></small>
      <div class="d-none name"><code>diffAnalyses.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Perform statistical analyses</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">diffAnalyses</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  groups <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  analyses <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"wilcoxRankSum"</span>, <span class="st">"ttest"</span>, <span class="st">"kruskal"</span>, <span class="st">"levene"</span>, <span class="st">"fligner"</span><span class="op">)</span>,</span>
<span>  pvalueAdjust <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  geneExpr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  inputID <span class="op">=</span> <span class="st">"sparklineInput"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>data</dt>
<dd><p>Data frame or matrix: gene expression or alternative splicing
quantification</p></dd>


<dt>groups</dt>
<dd><p>Named list of characters (containing elements belonging to each
group) or character vector (containing the group of each individual sample);
if <code>NULL</code>, sample types are used instead when available, e.g. normal,
tumour and metastasis</p></dd>


<dt>analyses</dt>
<dd><p>Character: statistical tests to perform (see Details)</p></dd>


<dt>pvalueAdjust</dt>
<dd><p>Character: method used to adjust p-values (see Details)</p></dd>


<dt>geneExpr</dt>
<dd><p>Character: name of the gene expression dataset (only required
for density sparklines available in the interactive mode)</p></dd>


<dt>inputID</dt>
<dd><p>Character: identifier of input to get attributes of clicked
event (Shiny only)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Table of statistical analyses</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The following statistical analyses may be performed simultaneously via the
<code>analysis</code> argument:</p><ul><li><p><code>ttest</code> - Unpaired t-test (2 groups)</p></li>
<li><p><code>wilcoxRankSum</code> - Wilcoxon Rank Sum test (2 groups)</p></li>
<li><p><code>kruskal</code> - Kruskal test (2 or more groups)</p></li>
<li><p><code>levene</code> - Levene's test (2 or more groups)</p></li>
<li><p><code>fligner</code> - Fligner-Killeen test (2 or more groups)</p></li>
<li><p><code>density</code> - Sample distribution per group (only usable
     through the visual interface)</p></li>
</ul><p>The following p-value adjustment methods are supported via the
<code>pvalueAdjust</code> argument:</p><ul><li><p><code>none</code>: do not adjust p-values</p></li>
<li><p><code>BH</code>: Benjamini-Hochberg's method (false discovery rate)</p></li>
<li><p><code>BY</code>: Benjamini-Yekutieli's method (false discovery rate)</p></li>
<li><p><code>bonferroni</code>: Bonferroni correction (family-wise error rate)</p></li>
<li><p><code>holm</code>: Holm's method (family-wise error rate)</p></li>
<li><p><code>hochberg</code>: Hochberg's method (family-wise error rate)</p></li>
<li><p><code>hommel</code>: Hommel's method (family-wise error rate)</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other functions to perform and plot differential analyses: 
<code><a href="plotDistribution.html">plotDistribution</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Calculate PSI for skipped exon (SE) and mutually exclusive (MXE) events</span></span></span>
<span class="r-in"><span><span class="va">eventType</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SE"</span>, <span class="st">"MXE"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="readFile.html">readFile</a></span><span class="op">(</span><span class="st">"ex_splicing_annotation.RDS"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">junctionQuant</span> <span class="op">&lt;-</span> <span class="fu"><a href="readFile.html">readFile</a></span><span class="op">(</span><span class="st">"ex_junctionQuant.RDS"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">psi</span> <span class="op">&lt;-</span> <span class="fu"><a href="quantifySplicing.html">quantifySplicing</a></span><span class="op">(</span><span class="va">annot</span>, <span class="va">junctionQuant</span>, eventType<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SE"</span>, <span class="st">"MXE"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using 3 of 3 events (100%) whose junctions are present in junction quantification data...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                        |   0%   |========                                |  20%   |================                        |  40%   |========================                |  60%   |================================        |  80%   |========================================| 100% </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using 3 of 3 events (100%) whose junctions are present in junction quantification data...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                        |   0%   |========                                |  20%   |================                        |  40%   |========================                |  60%   |================================        |  80%   |========================================| 100% </span>
<span class="r-in"><span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Normal"</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Tumour"</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">diffAnalyses</span><span class="op">(</span><span class="va">psi</span>, <span class="va">group</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Time difference of 0.0368 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Time difference of 0.00995 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Time difference of 0.000438 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Time difference of 0.00147 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                    Event type Chromosome Strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                    Skipped exon (SE)          1      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1 Mutually exclusive exon (MXE)          1      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                    Skipped exon (SE)          2      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                    Skipped exon (SE)          3      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA Mutually exclusive exon (MXE)          2      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6 Mutually exclusive exon (MXE)          3      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                  Gene Survival by PSI cutoff Optimal PSI cutoff</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1        ACTN1                   &lt;NA&gt;               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1 ACTN1                   &lt;NA&gt;               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA        ALDOA                   &lt;NA&gt;               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6        ANXA6                   &lt;NA&gt;               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA ALDOA                   &lt;NA&gt;               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6 ANXA6                   &lt;NA&gt;               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Log-rank p-value Samples (Normal)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                    &lt;NA&gt;                3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1             &lt;NA&gt;                3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                    &lt;NA&gt;                3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                    &lt;NA&gt;                3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA             &lt;NA&gt;                3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6             &lt;NA&gt;                3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Samples (Tumour) Wilcoxon statistic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                       3                4.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1                3                4.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                       3                6.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                       3                4.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA                3                8.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6                3                2.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Wilcoxon p-value Wilcoxon p-value (BH adjusted)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                     NaN                            NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1              NaN                            NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                     0.7                      0.9333333</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                     1.0                      1.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA              0.2                      0.8000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6              0.4                      0.8000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Wilcoxon null value Wilcoxon alternative</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                          0            two.sided</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1                   0            two.sided</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                          0            two.sided</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                          0            two.sided</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA                   0            two.sided</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6                   0            two.sided</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                                   Wilcoxon method</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1        Wilcoxon rank sum test with continuity correction</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1 Wilcoxon rank sum test with continuity correction</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                             Wilcoxon rank sum exact test</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                             Wilcoxon rank sum exact test</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA                      Wilcoxon rank sum exact test</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6                      Wilcoxon rank sum exact test</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                   Wilcoxon data name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1        vector[typeOne] and vector[!typeOne]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1 vector[typeOne] and vector[!typeOne]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA        vector[typeOne] and vector[!typeOne]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6        vector[typeOne] and vector[!typeOne]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA vector[typeOne] and vector[!typeOne]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6 vector[typeOne] and vector[!typeOne]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Kruskal statistic Kruskal parameter</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                      NaN                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1               NaN                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA               0.42857143                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6               0.04761905                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA        2.33333333                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6        1.19047619                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Kruskal p-value Kruskal p-value (BH adjusted)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                    NaN                           NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1             NaN                           NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA              0.5126908                     0.6835877</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6              0.8272593                     0.8272593</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA       0.1266305                     0.5065218</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6       0.2752335                     0.5504670</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                               Kruskal method Kruskal data name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1        Kruskal-Wallis rank sum test  vector and group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1 Kruskal-Wallis rank sum test  vector and group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA        Kruskal-Wallis rank sum test  vector and group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6        Kruskal-Wallis rank sum test  vector and group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA Kruskal-Wallis rank sum test  vector and group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6 Kruskal-Wallis rank sum test  vector and group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Levene statistic Levene p-value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                     NaN            NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1              NaN            NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                1.534423      0.2831735</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                1.898555      0.2403023</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA         1.102360      0.3529948</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6         1.453025      0.2944743</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Levene p-value (BH adjusted) Levene data name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                                 NaN vector and group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1                          NaN vector and group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                           0.3529948 vector and group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                           0.3529948 vector and group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA                    0.3529948 vector and group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6                    0.3529948 vector and group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                    Levene method</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1        Levene's test (using the median)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1 Levene's test (using the median)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA        Levene's test (using the median)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6        Levene's test (using the median)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA Levene's test (using the median)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6 Levene's test (using the median)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Fligner-Killeen statistic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                              NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1                       NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                         1.042057</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                         1.042057</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA                  1.042057</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6                  1.042057</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Fligner-Killeen parameter</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                                1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1                         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                                1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                                1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA                         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6                         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Fligner-Killeen p-value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                            NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1                     NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                      0.3073435</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                      0.3073435</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA               0.3073435</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6               0.3073435</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Fligner-Killeen p-value (BH adjusted)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                                          NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1                                   NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                                    0.3073435</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                                    0.3073435</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA                             0.3073435</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6                             0.3073435</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                           Fligner-Killeen method</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1        Fligner-Killeen test of homogeneity of variances</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1 Fligner-Killeen test of homogeneity of variances</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA        Fligner-Killeen test of homogeneity of variances</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6        Fligner-Killeen test of homogeneity of variances</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA Fligner-Killeen test of homogeneity of variances</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6 Fligner-Killeen test of homogeneity of variances</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Fligner-Killeen data name Variance (Normal)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                 vector and group       0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1          vector and group       0.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                 vector and group       0.026663118</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                 vector and group       0.005240524</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA          vector and group       0.019685695</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6          vector and group       0.022583581</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Variance (Tumour) Median (Normal)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1             0.0000000000       0.5000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1      0.0000000000       0.5000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA             0.0027119499       0.7402597</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6             0.0003158887       0.4255319</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA      0.0035641399       0.7500000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6      0.0027577471       0.2597403</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Median (Tumour) T-test statistic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1              0.5000000               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1       0.5000000               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA              0.6037736       1.04878918</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6              0.4418605      -0.02602006</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA       0.5918367       1.73393229</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6       0.3846154      -1.45806393</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 T-test parameter T-test p-value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                      NA             NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1               NA             NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                2.402681      0.3881661</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                2.240239      0.9813765</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA         2.701223      0.1913725</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6         2.481275      0.2586925</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 T-test p-value (BH adjusted) T-test conf int1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                                  NA               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1                           NA               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA                           0.5175548       -0.2604062</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6                           0.9813765       -0.1685048</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA                    0.5173850       -0.1458915</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6                    0.5173850       -0.4643286</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 T-test conf int2 T-test estimate1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                      NA               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1               NA               NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA               0.4679678        0.7217532</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6               0.1662652        0.4353827</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA        0.4511807        0.7434609</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6        0.1963126        0.2671886</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 T-test estimate2 T-test null value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                      NA                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1               NA                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA               0.6179724                 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6               0.4365025                 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA        0.5908163                 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6        0.4011966                 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 T-test stderr T-test alternative</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                   NA               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1            NA               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA           0.09895296          two.sided</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6           0.04303647          two.sided</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA    0.08803377          two.sided</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6    0.09190816          two.sided</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                           T-test method</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                           &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1                    &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA        Welch Two Sample t-test</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6        Welch Two Sample t-test</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA Welch Two Sample t-test</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6 Welch Two Sample t-test</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                     T-test data name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1                                        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1                                 &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA        vector[typeOne] and vector[!typeOne]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6        vector[typeOne] and vector[!typeOne]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA vector[typeOne] and vector[!typeOne]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6 vector[typeOne] and vector[!typeOne]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                  ∆ Variance    ∆ Median</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_1_+_32_35_37_38_ACTN1        0.000000000  0.00000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_1_+_32_35_37_38_40_42_ACTN1 0.000000000  0.00000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_2_+_32_35_37_38_ALDOA        0.023951168  0.13648616</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SE_3_+_32_35_37_38_ANXA6        0.004924635 -0.01632855</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_2_+_32_35_37_38_40_42_ALDOA 0.016121555  0.15816327</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MXE_3_+_32_35_37_38_40_42_ANXA6 0.019825834 -0.12487512</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

