<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference • psichomics</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script>


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">psichomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.12.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    </li>
    <li>
      <a href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    </li>
    <li>
      <a href="../articles/custom_data.html">Loading SRA and user-provided RNA-seq data</a>
    </li>
    <li>
      <a href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-general-functions" class="hasAnchor"><a href="#section-general-functions" class="anchor"></a>General functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="psichomics.html">psichomics()</a></code> </p>
        </td>
        <td><p>Start graphical interface of psichomics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parseSplicingEvent.html">parseSplicingEvent()</a></code> </p>
        </td>
        <td><p>Parse alternative splicing event identifier</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getSplicingEventFromGenes.html">getSplicingEventFromGenes()</a></code> <code><a href="getSplicingEventFromGenes.html">getGenesFromSplicingEvents()</a></code> </p>
        </td>
        <td><p>Get alternative splicing events from genes or vice-versa</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSplicingEvent.html">plotSplicingEvent()</a></code> </p>
        </td>
        <td><p>Plot diagram of alternative splicing events</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDownloadsFolder.html">getDownloadsFolder()</a></code> </p>
        </td>
        <td><p>Get the path to the Downloads folder</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterGroups.html">filterGroups()</a></code> </p>
        </td>
        <td><p>Filter groups with less data points than the threshold</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-tcga-data-retrieval" class="hasAnchor"><a href="#section-tcga-data-retrieval" class="anchor"></a>TCGA data retrieval</h2>
          <p class="section-desc"><p>Retrieve TCGA data using <a href="http://firebrowse.org/api-docs/">Firebrowse API</a></p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="isFirebrowseUp.html">isFirebrowseUp()</a></code> </p>
        </td>
        <td><p>Check if <a href='http://firebrowse.org/api-docs/'>Firebrowse API</a> is running</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getTCGAdataTypes.html">getTCGAdataTypes()</a></code> <code><a href="getTCGAdataTypes.html">getTCGAdates()</a></code> <code><a href="getTCGAdataTypes.html">getTCGAcohorts()</a></code> </p>
        </td>
        <td><p>Get available parameters for TCGA data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadTCGAdata.html">loadTCGAdata()</a></code> </p>
        </td>
        <td><p>Download and process TCGA data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parseTcgaSampleInfo.html">parseTCGAsampleTypes()</a></code> <code><a href="parseTcgaSampleInfo.html">parseTCGAsampleInfo()</a></code> </p>
        </td>
        <td><p>Parse sample information from TCGA sample identifiers</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-gtex-data-retrieval" class="hasAnchor"><a href="#section-gtex-data-retrieval" class="anchor"></a>GTEx data retrieval</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="getGtexDataTypes.html">getGtexDataTypes()</a></code> </p>
        </td>
        <td><p>Get GTEx data types</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGtexTissues.html">getGtexTissues()</a></code> </p>
        </td>
        <td><p>Get GTEx tissues from given GTEx sample attributes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadGtexData.html">loadGtexData()</a></code> </p>
        </td>
        <td><p>Load GTEx data</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-sra-data-retrieval" class="hasAnchor"><a href="#section-sra-data-retrieval" class="anchor"></a>SRA data retrieval</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="loadSRAproject.html">loadSRAproject()</a></code> </p>
        </td>
        <td><p>Download and load SRA projects via
<a href='https://jhubiostatistics.shinyapps.io/recount/'>recount2</a></p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-user-provided-data-loading" class="hasAnchor"><a href="#section-user-provided-data-loading" class="anchor"></a>User-provided data loading</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="loadLocalFiles.html">loadLocalFiles()</a></code> </p>
        </td>
        <td><p>Load local files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepareSRAmetadata.html">prepareSRAmetadata()</a></code> <code><a href="prepareSRAmetadata.html">prepareJunctionQuant()</a></code> <code><a href="prepareSRAmetadata.html">prepareGeneQuant()</a></code> </p>
        </td>
        <td><p>Prepare user-provided files to be loaded into psichomics</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-gene-expression-pre-processing" class="hasAnchor"><a href="#section-gene-expression-pre-processing" class="anchor"></a>Gene expression pre-processing</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="plotRowStats.html">plotRowStats()</a></code> </p>
        </td>
        <td><p>Plot sample statistics per row</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGeneExprPerSample.html">plotGeneExprPerSample()</a></code> </p>
        </td>
        <td><p>Plot distribution of gene expression per sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterGeneExpr.html">filterGeneExpr()</a></code> </p>
        </td>
        <td><p>Filter genes based on their expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="normaliseGeneExpression.html">normaliseGeneExpression()</a></code> </p>
        </td>
        <td><p>Filter and normalise gene expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertGeneIdentifiers.html">convertGeneIdentifiers()</a></code> </p>
        </td>
        <td><p>Convert gene identifiers</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-psi-quantification" class="hasAnchor"><a href="#section-psi-quantification" class="anchor"></a>PSI quantification</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="getSplicingEventTypes.html">getSplicingEventTypes()</a></code> </p>
        </td>
        <td><p>Get supported splicing event types</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="listSplicingAnnotations.html">listSplicingAnnotations()</a></code> </p>
        </td>
        <td><p>List alternative splicing annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadAnnotation.html">loadAnnotation()</a></code> </p>
        </td>
        <td><p>Load alternative splicing annotation from <code>AnnotationHub</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="quantifySplicing.html">quantifySplicing()</a></code> </p>
        </td>
        <td><p>Quantify alternative splicing events</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotRowStats.html">plotRowStats()</a></code> </p>
        </td>
        <td><p>Plot sample statistics per row</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterPSI.html">filterPSI()</a></code> </p>
        </td>
        <td><p>Filter alternative splicing quantification</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-custom-alternative-splicing-annotation-preparation" class="hasAnchor"><a href="#section-custom-alternative-splicing-annotation-preparation" class="anchor"></a>Custom alternative splicing annotation preparation</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="prepareAnnotationFromEvents.html">prepareAnnotationFromEvents()</a></code> </p>
        </td>
        <td><p>Prepare annotation from alternative splicing events</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parseMisoAnnotation.html">parseSuppaAnnotation()</a></code> <code><a href="parseMisoAnnotation.html">parseVastToolsAnnotation()</a></code> <code><a href="parseMisoAnnotation.html">parseMisoAnnotation()</a></code> <code><a href="parseMisoAnnotation.html">parseMatsAnnotation()</a></code> </p>
        </td>
        <td><p>Get events from alternative splicing annotation</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-data-grouping" class="hasAnchor"><a href="#section-data-grouping" class="anchor"></a>Data Grouping</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="getGeneList.html">getGeneList()</a></code> </p>
        </td>
        <td><p>Get pre-created, literature-based gene list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createGroupByAttribute.html">createGroupByAttribute()</a></code> </p>
        </td>
        <td><p>Split elements into groups based on a given column of a dataset</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getMatchingSamples.html">getMatchingSamples()</a></code> </p>
        </td>
        <td><p>Get samples matching the given subjects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getSubjectFromSample.html">getSubjectFromSample()</a></code> </p>
        </td>
        <td><p>Get subjects from given samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="groupPerElem.html">groupPerElem()</a></code> </p>
        </td>
        <td><p>Assign one group to each element</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="testGroupIndependence.html">testGroupIndependence()</a></code> </p>
        </td>
        <td><p>Multiple independence tests between reference groups and list of groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGroupIndependence.html">plotGroupIndependence()</a></code> </p>
        </td>
        <td><p>Plot <code>-log10(p-values)</code> of the results obtained after multiple group
independence testing</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-principal-component-analysis-pca-" class="hasAnchor"><a href="#section-principal-component-analysis-pca-" class="anchor"></a>Principal component analysis (PCA)</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="performPCA.html">performPCA()</a></code> </p>
        </td>
        <td><p>Perform principal component analysis after processing missing values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotVariance.html">plotVariance()</a></code> </p>
        </td>
        <td><p>Create the explained variance plot from a PCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculateLoadingsContribution.html">calculateLoadingsContribution()</a></code> </p>
        </td>
        <td><p>Calculate the contribution of PCA loadings to the selected principal
components</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotPCA.html">plotPCA()</a></code> </p>
        </td>
        <td><p>Create a scatterplot from a PCA object</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-independent-component-analysis-ica-" class="hasAnchor"><a href="#section-independent-component-analysis-ica-" class="anchor"></a>Independent component analysis (ICA)</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="performICA.html">performICA()</a></code> </p>
        </td>
        <td><p>Perform independent component analysis after processing missing values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotICA.html">plotICA()</a></code> </p>
        </td>
        <td><p>Create multiple scatterplots from ICA</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-survival-analysis" class="hasAnchor"><a href="#section-survival-analysis" class="anchor"></a>Survival analysis</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="getAttributesTime.html">getAttributesTime()</a></code> </p>
        </td>
        <td><p>Get time values for given columns in a clinical dataset</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assignValuePerSubject.html">assignValuePerSubject()</a></code> </p>
        </td>
        <td><p>Assign average sample values to their corresponding subjects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="labelBasedOnCutoff.html">labelBasedOnCutoff()</a></code> </p>
        </td>
        <td><p>Label groups based on a given cutoff</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="optimalSurvivalCutoff.html">optimalSurvivalCutoff()</a></code> </p>
        </td>
        <td><p>Calculate optimal data cutoff that best separates survival curves</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="testSurvival.html">testSurvival()</a></code> </p>
        </td>
        <td><p>Test the survival difference between groups of subjects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="processSurvTerms.html">processSurvTerms()</a></code> </p>
        </td>
        <td><p>Process survival curves terms to calculate survival curves</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="survfit.survTerms.html">survfit(<i>&lt;survTerms&gt;</i>)</a></code> </p>
        </td>
        <td><p>Create survival curves</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="survdiffTerms.html">survdiffTerms()</a></code> </p>
        </td>
        <td><p>Test Survival Curve Differences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSurvivalCurves.html">plotSurvivalCurves()</a></code> </p>
        </td>
        <td><p>Plot survival curves</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSurvivalPvaluesByCutoff.html">plotSurvivalPvaluesByCutoff()</a></code> </p>
        </td>
        <td><p>Plot p-values of survival difference between groups based on multiple cutoffs</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-differential-analyses" class="hasAnchor"><a href="#section-differential-analyses" class="anchor"></a>Differential analyses</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="diffAnalyses.html">diffAnalyses()</a></code> </p>
        </td>
        <td><p>Perform statistical analyses</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDistribution.html">plotDistribution()</a></code> </p>
        </td>
        <td><p>Plot distribution through a density plot</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-gene-expression-and-alternative-splicing-correlation" class="hasAnchor"><a href="#section-gene-expression-and-alternative-splicing-correlation" class="anchor"></a>Gene expression and alternative splicing correlation</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="correlateGEandAS.html">correlateGEandAS()</a></code> </p>
        </td>
        <td><p>Correlate gene expression data against alternative splicing quantification</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot.GEandAScorrelation.html">`[`(<i>&lt;GEandAScorrelation&gt;</i>)</a></code> <code><a href="plot.GEandAScorrelation.html">plot(<i>&lt;GEandAScorrelation&gt;</i>)</a></code> <code><a href="plot.GEandAScorrelation.html">print(<i>&lt;GEandAScorrelation&gt;</i>)</a></code> <code><a href="plot.GEandAScorrelation.html">as.table(<i>&lt;GEandAScorrelation&gt;</i>)</a></code> </p>
        </td>
        <td><p>Display results of correlation analyses</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-gene-transcript-and-protein-annotation-retrieval" class="hasAnchor"><a href="#section-gene-transcript-and-protein-annotation-retrieval" class="anchor"></a>Gene, transcript and protein annotation retrieval</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="queryEnsemblByGene.html">queryEnsemblByGene()</a></code> <code><a href="queryEnsemblByGene.html">queryEnsemblByEvent()</a></code> </p>
        </td>
        <td><p>Query information from Ensembl</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ensemblToUniprot.html">ensemblToUniprot()</a></code> </p>
        </td>
        <td><p>Convert from Ensembl to UniProt identifier</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotProtein.html">plotProtein()</a></code> </p>
        </td>
        <td><p>Plot protein features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotTranscripts.html">plotTranscripts()</a></code> </p>
        </td>
        <td><p>Plot transcripts</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-general-functions">General functions</a></li>
      <li><a href="#section-tcga-data-retrieval">TCGA data retrieval</a></li>
      <li><a href="#section-gtex-data-retrieval">GTEx data retrieval</a></li>
      <li><a href="#section-sra-data-retrieval">SRA data retrieval</a></li>
      <li><a href="#section-user-provided-data-loading">User-provided data loading</a></li>
      <li><a href="#section-gene-expression-pre-processing">Gene expression pre-processing</a></li>
      <li><a href="#section-psi-quantification">PSI quantification</a></li>
      <li><a href="#section-custom-alternative-splicing-annotation-preparation">Custom alternative splicing annotation preparation</a></li>
      <li><a href="#section-data-grouping">Data Grouping</a></li>
      <li><a href="#section-principal-component-analysis-pca-">Principal component analysis (PCA)</a></li>
      <li><a href="#section-independent-component-analysis-ica-">Independent component analysis (ICA)</a></li>
      <li><a href="#section-survival-analysis">Survival analysis</a></li>
      <li><a href="#section-differential-analyses">Differential analyses</a></li>
      <li><a href="#section-gene-expression-and-alternative-splicing-correlation">Gene expression and alternative splicing correlation</a></li>
      <li><a href="#section-gene-transcript-and-protein-annotation-retrieval">Gene, transcript and protein annotation retrieval</a></li>
    </ul>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


