<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference • psichomics</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">psichomics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.20.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    <a class="dropdown-item" href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    <a class="dropdown-item" href="../articles/custom_data.html">Loading user-provided data</a>
    <a class="dropdown-item" href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nuno-agostinho/psichomics/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="general-functions">General functions<a class="anchor" aria-label="anchor" href="#general-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="psichomics.html">psichomics()</a></code> 
        </dt>
        <dd>Start graphical interface of psichomics</dd>
      </dl><dl><dt>
          
          <code><a href="parseSplicingEvent.html">parseSplicingEvent()</a></code> 
        </dt>
        <dd>Parse alternative splicing event identifier</dd>
      </dl><dl><dt>
          
          <code><a href="getSplicingEventData.html">getSplicingEventData()</a></code> 
        </dt>
        <dd>Get splicing event information for given alternative splicing quantification
data</dd>
      </dl><dl><dt>
          
          <code><a href="getSplicingEventFromGenes.html">getSplicingEventFromGenes()</a></code> <code><a href="getSplicingEventFromGenes.html">getGenesFromSplicingEvents()</a></code> 
        </dt>
        <dd>Get alternative splicing events from genes or vice-versa</dd>
      </dl><dl><dt>
          
          <code><a href="plotSplicingEvent.html">plotSplicingEvent()</a></code> 
        </dt>
        <dd>Plot diagram of alternative splicing events</dd>
      </dl><dl><dt>
          
          <code><a href="parseCategoricalGroups.html">parseCategoricalGroups()</a></code> 
        </dt>
        <dd>Parse categorical columns in a data frame</dd>
      </dl><dl><dt>
          
          <code><a href="getDownloadsFolder.html">getDownloadsFolder()</a></code> 
        </dt>
        <dd>Get the path to the Downloads folder</dd>
      </dl></div><div class="section level2">
      <h2 id="tcga-data-retrieval">TCGA data retrieval<a class="anchor" aria-label="anchor" href="#tcga-data-retrieval"></a></h2>
      
      <p class="section-desc"></p><p>Retrieve TCGA data using <a href="http://firebrowse.org/api-docs/" class="external-link">Firebrowse API</a></p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="isFirebrowseUp.html">isFirebrowseUp()</a></code> 
        </dt>
        <dd>Check if <a href="http://firebrowse.org/api-docs/" class="external-link">FireBrowse API</a> is running</dd>
      </dl><dl><dt>
          
          <code><a href="getTCGAdataTypes.html">getTCGAdataTypes()</a></code> <code><a href="getTCGAdataTypes.html">getTCGAdates()</a></code> <code><a href="getTCGAdataTypes.html">getTCGAcohorts()</a></code> 
        </dt>
        <dd>Get available parameters for TCGA data</dd>
      </dl><dl><dt>
          
          <code><a href="loadTCGAdata.html">loadTCGAdata()</a></code> 
        </dt>
        <dd>Download and process TCGA data</dd>
      </dl><dl><dt>
          
          <code><a href="parseTcgaSampleInfo.html">parseTCGAsampleTypes()</a></code> <code><a href="parseTcgaSampleInfo.html">parseTCGAsampleInfo()</a></code> 
        </dt>
        <dd>Parse sample information from TCGA sample identifiers</dd>
      </dl></div><div class="section level2">
      <h2 id="gtex-data-retrieval">GTEx data retrieval<a class="anchor" aria-label="anchor" href="#gtex-data-retrieval"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="getGtexDataTypes.html">getGtexDataTypes()</a></code> <code><a href="getGtexDataTypes.html">getGtexReleases()</a></code> 
        </dt>
        <dd>Get GTEx data information</dd>
      </dl><dl><dt>
          
          <code><a href="getGtexTissues.html">getGtexTissues()</a></code> 
        </dt>
        <dd>Get GTEx tissues from given GTEx sample attributes</dd>
      </dl><dl><dt>
          
          <code><a href="loadGtexData.html">loadGtexData()</a></code> 
        </dt>
        <dd>Download and load GTEx data</dd>
      </dl></div><div class="section level2">
      <h2 id="sra-data-retrieval">SRA data retrieval<a class="anchor" aria-label="anchor" href="#sra-data-retrieval"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="loadSRAproject.html">loadSRAproject()</a></code> 
        </dt>
        <dd>Download and load SRA projects via
<a href="https://jhubiostatistics.shinyapps.io/recount/" class="external-link">recount2</a></dd>
      </dl></div><div class="section level2">
      <h2 id="user-provided-data-loading">User-provided data loading<a class="anchor" aria-label="anchor" href="#user-provided-data-loading"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="loadLocalFiles.html">loadLocalFiles()</a></code> 
        </dt>
        <dd>Load local files</dd>
      </dl><dl><dt>
          
          <code><a href="prepareSRAmetadata.html">prepareSRAmetadata()</a></code> <code><a href="prepareSRAmetadata.html">prepareJunctionQuant()</a></code> <code><a href="prepareSRAmetadata.html">prepareGeneQuant()</a></code> 
        </dt>
        <dd>Prepare user-provided files to be loaded into psichomics</dd>
      </dl></div><div class="section level2">
      <h2 id="gene-expression-pre-processing">Gene expression pre-processing<a class="anchor" aria-label="anchor" href="#gene-expression-pre-processing"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="plotRowStats.html">plotRowStats()</a></code> 
        </dt>
        <dd>Plot row-wise statistics</dd>
      </dl><dl><dt>
          
          <code><a href="plotGeneExprPerSample.html">plotGeneExprPerSample()</a></code> 
        </dt>
        <dd>Plot distribution of gene expression per sample</dd>
      </dl><dl><dt>
          
          <code><a href="plotLibrarySize.html">plotLibrarySize()</a></code> 
        </dt>
        <dd>Plot library size</dd>
      </dl><dl><dt>
          
          <code><a href="filterGeneExpr.html">filterGeneExpr()</a></code> 
        </dt>
        <dd>Filter genes based on their expression</dd>
      </dl><dl><dt>
          
          <code><a href="normaliseGeneExpression.html">normaliseGeneExpression()</a></code> <code><a href="normaliseGeneExpression.html">normalizeGeneExpression()</a></code> 
        </dt>
        <dd>Filter and normalise gene expression</dd>
      </dl><dl><dt>
          
          <code><a href="convertGeneIdentifiers.html">convertGeneIdentifiers()</a></code> 
        </dt>
        <dd>Convert gene identifiers</dd>
      </dl></div><div class="section level2">
      <h2 id="psi-quantification-and-filtering">PSI quantification and filtering<a class="anchor" aria-label="anchor" href="#psi-quantification-and-filtering"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="getSplicingEventTypes.html">getSplicingEventTypes()</a></code> 
        </dt>
        <dd>Get supported splicing event types</dd>
      </dl><dl><dt>
          
          <code><a href="listSplicingAnnotations.html">listSplicingAnnotations()</a></code> 
        </dt>
        <dd>List alternative splicing annotations</dd>
      </dl><dl><dt>
          
          <code><a href="loadAnnotation.html">loadAnnotation()</a></code> 
        </dt>
        <dd>Load alternative splicing annotation from <code>AnnotationHub</code></dd>
      </dl><dl><dt>
          
          <code><a href="quantifySplicing.html">quantifySplicing()</a></code> 
        </dt>
        <dd>Quantify alternative splicing events</dd>
      </dl><dl><dt>
          
          <code><a href="discardLowCoveragePSIvalues.html">discardLowCoveragePSIvalues()</a></code> 
        </dt>
        <dd>Remove alternative splicing quantification values based on coverage</dd>
      </dl><dl><dt>
          
          <code><a href="filterPSI.html">filterPSI()</a></code> 
        </dt>
        <dd>Filter alternative splicing quantification</dd>
      </dl><dl><dt>
          
          <code><a href="plotRowStats.html">plotRowStats()</a></code> 
        </dt>
        <dd>Plot row-wise statistics</dd>
      </dl></div><div class="section level2">
      <h2 id="custom-alternative-splicing-annotation">Custom alternative splicing annotation<a class="anchor" aria-label="anchor" href="#custom-alternative-splicing-annotation"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="parseMisoAnnotation.html">parseSuppaAnnotation()</a></code> <code><a href="parseMisoAnnotation.html">parseVastToolsAnnotation()</a></code> <code><a href="parseMisoAnnotation.html">parseMisoAnnotation()</a></code> <code><a href="parseMisoAnnotation.html">parseMatsAnnotation()</a></code> 
        </dt>
        <dd>Parse events from alternative splicing annotation</dd>
      </dl><dl><dt>
          
          <code><a href="prepareAnnotationFromEvents.html">prepareAnnotationFromEvents()</a></code> 
        </dt>
        <dd>Prepare annotation from alternative splicing events</dd>
      </dl></div><div class="section level2">
      <h2 id="data-grouping">Data Grouping<a class="anchor" aria-label="anchor" href="#data-grouping"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="createGroupByAttribute.html">createGroupByAttribute()</a></code> 
        </dt>
        <dd>Split elements into groups based on a given column of a dataset</dd>
      </dl><dl><dt>
          
          <code><a href="filterGroups.html">filterGroups()</a></code> 
        </dt>
        <dd>Filter groups with less data points than the threshold</dd>
      </dl><dl><dt>
          
          <code><a href="groupPerElem.html">groupPerElem()</a></code> 
        </dt>
        <dd>Assign one group to each element</dd>
      </dl><dl><dt>
          
          <code><a href="getGeneList.html">getGeneList()</a></code> 
        </dt>
        <dd>Get curated, literature-based gene lists</dd>
      </dl><dl><dt>
          
          <code><a href="getSampleFromSubject.html">getSampleFromSubject()</a></code> 
        </dt>
        <dd>Get samples matching the given subjects</dd>
      </dl><dl><dt>
          
          <code><a href="getSubjectFromSample.html">getSubjectFromSample()</a></code> 
        </dt>
        <dd>Get subjects from given samples</dd>
      </dl><dl><dt>
          
          <code><a href="testGroupIndependence.html">testGroupIndependence()</a></code> 
        </dt>
        <dd>Multiple independence tests between reference groups and list of groups</dd>
      </dl><dl><dt>
          
          <code><a href="plotGroupIndependence.html">plotGroupIndependence()</a></code> 
        </dt>
        <dd>Plot <code>-log10(p-values)</code> of the results obtained after multiple group
independence testing</dd>
      </dl></div><div class="section level2">
      <h2 id="principal-component-analysis-pca-">Principal component analysis (PCA)<a class="anchor" aria-label="anchor" href="#principal-component-analysis-pca-"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="performPCA.html">performPCA()</a></code> 
        </dt>
        <dd>Perform principal component analysis after processing missing values</dd>
      </dl><dl><dt>
          
          <code><a href="plotPCAvariance.html">plotPCAvariance()</a></code> 
        </dt>
        <dd>Create the explained variance plot from a PCA</dd>
      </dl><dl><dt>
          
          <code><a href="calculateLoadingsContribution.html">calculateLoadingsContribution()</a></code> 
        </dt>
        <dd>Calculate the contribution of PCA loadings to the selected principal
components</dd>
      </dl><dl><dt>
          
          <code><a href="plotPCA.html">plotPCA()</a></code> 
        </dt>
        <dd>Create a scatterplot from a PCA object</dd>
      </dl></div><div class="section level2">
      <h2 id="independent-component-analysis-ica-">Independent component analysis (ICA)<a class="anchor" aria-label="anchor" href="#independent-component-analysis-ica-"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="performICA.html">performICA()</a></code> 
        </dt>
        <dd>Perform independent component analysis after processing missing values</dd>
      </dl><dl><dt>
          
          <code><a href="plotICA.html">plotICA()</a></code> 
        </dt>
        <dd>Create multiple scatterplots from ICA</dd>
      </dl></div><div class="section level2">
      <h2 id="differential-analyses">Differential analyses<a class="anchor" aria-label="anchor" href="#differential-analyses"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="diffAnalyses.html">diffAnalyses()</a></code> 
        </dt>
        <dd>Perform statistical analyses</dd>
      </dl><dl><dt>
          
          <code><a href="plotDistribution.html">plotDistribution()</a></code> 
        </dt>
        <dd>Plot sample distribution</dd>
      </dl></div><div class="section level2">
      <h2 id="gene-expression-and-alternative-splicing-correlation">Gene expression and alternative splicing correlation<a class="anchor" aria-label="anchor" href="#gene-expression-and-alternative-splicing-correlation"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="correlateGEandAS.html">correlateGEandAS()</a></code> 
        </dt>
        <dd>Correlate gene expression data against alternative splicing quantification</dd>
      </dl><dl><dt>
          
          <code><a href="plot.GEandAScorrelation.html">`[`(<i>&lt;GEandAScorrelation&gt;</i>)</a></code> <code><a href="plot.GEandAScorrelation.html">plot(<i>&lt;GEandAScorrelation&gt;</i>)</a></code> <code><a href="plot.GEandAScorrelation.html">print(<i>&lt;GEandAScorrelation&gt;</i>)</a></code> <code><a href="plot.GEandAScorrelation.html">as.table(<i>&lt;GEandAScorrelation&gt;</i>)</a></code> 
        </dt>
        <dd>Display results of correlation analyses</dd>
      </dl></div><div class="section level2">
      <h2 id="survival-analysis">Survival analysis<a class="anchor" aria-label="anchor" href="#survival-analysis"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="getAttributesTime.html">getAttributesTime()</a></code> 
        </dt>
        <dd>Get time values for given columns in a clinical dataset</dd>
      </dl><dl><dt>
          
          <code><a href="processSurvTerms.html">processSurvTerms()</a></code> 
        </dt>
        <dd>Process survival curves terms to calculate survival curves</dd>
      </dl><dl><dt>
          
          <code><a href="survfit.survTerms.html">survfit(<i>&lt;survTerms&gt;</i>)</a></code> 
        </dt>
        <dd>Create survival curves</dd>
      </dl><dl><dt>
          
          <code><a href="survdiffTerms.html">survdiffTerms()</a></code> 
        </dt>
        <dd>Test Survival Curve Differences</dd>
      </dl><dl><dt>
          
          <code><a href="plotSurvivalCurves.html">plotSurvivalCurves()</a></code> 
        </dt>
        <dd>Plot survival curves</dd>
      </dl><dl><dt>
          
          <code><a href="testSurvival.html">testSurvival()</a></code> 
        </dt>
        <dd>Test the survival difference between groups of subjects</dd>
      </dl><dl><dt>
          
          <code><a href="assignValuePerSubject.html">assignValuePerSubject()</a></code> 
        </dt>
        <dd>Assign average sample values to their corresponding subjects</dd>
      </dl><dl><dt>
          
          <code><a href="labelBasedOnCutoff.html">labelBasedOnCutoff()</a></code> 
        </dt>
        <dd>Label groups based on a given cutoff</dd>
      </dl><dl><dt>
          
          <code><a href="optimalSurvivalCutoff.html">optimalSurvivalCutoff()</a></code> 
        </dt>
        <dd>Calculate optimal data cutoff that best separates survival curves</dd>
      </dl><dl><dt>
          
          <code><a href="plotSurvivalPvaluesByCutoff.html">plotSurvivalPvaluesByCutoff()</a></code> 
        </dt>
        <dd>Plot p-values of survival difference between groups based on multiple cutoffs</dd>
      </dl></div><div class="section level2">
      <h2 id="gene-transcript-and-protein-annotation-retrieval">Gene, transcript and protein annotation retrieval<a class="anchor" aria-label="anchor" href="#gene-transcript-and-protein-annotation-retrieval"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="queryEnsemblByGene.html">queryEnsemblByGene()</a></code> <code><a href="queryEnsemblByGene.html">queryEnsemblByEvent()</a></code> 
        </dt>
        <dd>Query information from Ensembl</dd>
      </dl><dl><dt>
          
          <code><a href="ensemblToUniprot.html">ensemblToUniprot()</a></code> 
        </dt>
        <dd>Convert from Ensembl to UniProt identifier</dd>
      </dl><dl><dt>
          
          <code><a href="plotProtein.html">plotProtein()</a></code> 
        </dt>
        <dd>Plot protein features</dd>
      </dl><dl><dt>
          
          <code><a href="plotTranscripts.html">plotTranscripts()</a></code> 
        </dt>
        <dd>Plot transcripts</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer></div>

  

  

  </body></html>

