<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Filter and normalise gene expression — normaliseGeneExpression • psichomics</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Filter and normalise gene expression — normaliseGeneExpression" />
<meta property="og:description" content="Gene expression is filtered and normalised in the following steps:

Filter gene expression;
Normalise gene expression with calcNormFactors;
If performVoom = FALSE, compute counts per million (CPM) using
cpm and log2-transform values if
log2transform = TRUE;
If performVoom = TRUE, use voom to compute
log2-CPM, quantile-normalise (if method = &quot;quantile&quot;) and estimate
mean-variance relationship to calculate observation-level weights.

" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">psichomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.18.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    </li>
    <li>
      <a href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    </li>
    <li>
      <a href="../articles/custom_data.html">Loading user-provided data</a>
    </li>
    <li>
      <a href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/nuno-agostinho/psichomics/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Filter and normalise gene expression</h1>
    <small class="dont-index">Source: <a href='https://github.com/nuno-agostinho/psichomics/blob/master/R/data_geNormalisationFiltering.R'><code>R/data_geNormalisationFiltering.R</code></a></small>
    <div class="hidden name"><code>normaliseGeneExpression.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Gene expression is filtered and normalised in the following steps:</p>
<ul>
<li><p>Filter gene expression;</p></li>
<li><p>Normalise gene expression with <code><a href='https://rdrr.io/pkg/edgeR/man/calcNormFactors.html'>calcNormFactors</a></code>;</p></li>
<li><p>If <code>performVoom = FALSE</code>, compute counts per million (CPM) using
<code><a href='https://rdrr.io/pkg/edgeR/man/cpm.html'>cpm</a></code> and log2-transform values if
<code>log2transform = TRUE</code>;</p></li>
<li><p>If <code>performVoom = TRUE</code>, use <code><a href='https://rdrr.io/pkg/limma/man/voom.html'>voom</a></code> to compute
log2-CPM, quantile-normalise (if <code>method = "quantile"</code>) and estimate
mean-variance relationship to calculate observation-level weights.</p></li>
</ul>

    </div>

    <pre class="usage"><span class='fu'>normaliseGeneExpression</span><span class='op'>(</span>
  <span class='va'>geneExpr</span>,
  geneFilter <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  p <span class='op'>=</span> <span class='fl'>0.75</span>,
  log2transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  priorCount <span class='op'>=</span> <span class='fl'>0.25</span>,
  performVoom <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span>

<span class='fu'>normalizeGeneExpression</span><span class='op'>(</span>
  <span class='va'>geneExpr</span>,
  geneFilter <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  p <span class='op'>=</span> <span class='fl'>0.75</span>,
  log2transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  priorCount <span class='op'>=</span> <span class='fl'>0.25</span>,
  performVoom <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>geneExpr</th>
      <td><p>Matrix or data frame: gene expression</p></td>
    </tr>
    <tr>
      <th>geneFilter</th>
      <td><p>Boolean: filtered genes (if <code>NULL</code>, skip filtering)</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>Character: normalisation method, including <code>TMM</code>,
<code>RLE</code>, <code>upperquartile</code>, <code>none</code> or <code>quantile</code> (see
Details)</p></td>
    </tr>
    <tr>
      <th>p</th>
      <td><p>numeric value between 0 and 1 specifying which quantile of the counts should be used by <code>method="upperquartile"</code>.</p></td>
    </tr>
    <tr>
      <th>log2transform</th>
      <td><p>Boolean: perform log2-transformation?</p></td>
    </tr>
    <tr>
      <th>priorCount</th>
      <td><p>Average count to add to each observation to avoid zeroes
after log-transformation</p></td>
    </tr>
    <tr>
      <th>performVoom</th>
      <td><p>Boolean: perform mean-variance modelling
(using <code><a href='https://rdrr.io/pkg/limma/man/voom.html'>voom</a></code>)?</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Filtered and normalised gene expression</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code><a href='https://rdrr.io/pkg/edgeR/man/calcNormFactors.html'>edgeR::calcNormFactors</a></code> will be used to normalise gene
expression if <code>method</code> is <code>TMM</code>, <code>RLE</code>, <code>upperquartile</code>
or <code>none</code>. If <code>performVoom = TRUE</code>, <code><a href='https://rdrr.io/pkg/limma/man/voom.html'>voom</a></code> will
only normalise if <code>method = "quantile"</code>.</p>
<p>Available normalisation methods:</p><ul>
<li><p><code>TMM</code> is recommended for most RNA-seq data where more than half of
the genes are believed not differentially expressed between any pair of
samples;</p></li>
<li><p><code>RLE</code> calculates the median library from the geometric mean of all
columns and the median ratio of each sample to the median library is taken as
the scale factor;</p></li>
<li><p><code>upperquartile</code> calculates the scale factors from a given quantile
of the counts for each library, after removing genes with zero counts in all
libraries;</p></li>
<li><p><code>quantile</code> forces the entire empirical distribution of each
column to be identical (only performed if <code>performVoom = TRUE</code>).</p></li>
</ul>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other functions for gene expression pre-processing: 
<code><a href='convertGeneIdentifiers.html'>convertGeneIdentifiers</a>()</code>,
<code><a href='filterGeneExpr.html'>filterGeneExpr</a>()</code>,
<code><a href='plotGeneExprPerSample.html'>plotGeneExprPerSample</a>()</code>,
<code><a href='plotLibrarySize.html'>plotLibrarySize</a>()</code>,
<code><a href='plotRowStats.html'>plotRowStats</a>()</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='va'>geneExpr</span> <span class='op'>&lt;-</span> <span class='fu'><a href='readFile.html'>readFile</a></span><span class='op'>(</span><span class='st'>"ex_gene_expression.RDS"</span><span class='op'>)</span>
<span class='fu'>normaliseGeneExpression</span><span class='op'>(</span><span class='va'>geneExpr</span><span class='op'>)</span>
</div><div class='output co'>#&gt;                Normal 1 Normal 2 Normal 3 Normal 4 Normal 5 Cancer 1 Cancer 2
#&gt; ACTN1|87       16.44383 16.78304 16.48223 17.24207 17.07503 17.10327 16.49269
#&gt; ADAM15|8751    16.70099 16.09395 16.28898 15.96189 15.31259 15.63044 15.97064
#&gt; AKAP8L|26993   14.14424 14.80911 14.24193 14.41006 15.30310 14.58999 14.11916
#&gt; AKR1A1|10327   15.14212 15.06161 14.85207 15.74210 15.51387 15.57613 15.64140
#&gt; ALDOA|226      18.58900 20.00823 18.57857 18.85194 18.54547 18.33072 18.78489
#&gt; ANXA6|309      16.80155 16.85176 16.67479 16.76024 16.67841 16.79276 16.81781
#&gt; ARAP1|116985   16.01864 15.57970 15.45777 15.44823 15.82390 15.95734 15.89010
#&gt; ATP5C1|509     14.74478 15.47092 16.10871 15.23351 15.62633 15.44610 15.49723
#&gt; BOLA2|552900   14.03579 15.05401 14.67957 13.53149 14.32091 14.37295 14.08933
#&gt; C16orf13|84326 13.75582 14.21114 14.03332 13.69816 14.23023 13.78593 13.60679
#&gt;                Cancer 3 Cancer 4 Cancer 5
#&gt; ACTN1|87       16.81314 18.48316 17.61675
#&gt; ADAM15|8751    16.39172 15.97958 16.13043
#&gt; AKAP8L|26993   15.14118 14.98893 13.49192
#&gt; AKR1A1|10327   15.32114 15.90415 14.74053
#&gt; ALDOA|226      18.40211 18.50496 18.54164
#&gt; ANXA6|309      17.12530 16.53170 17.37841
#&gt; ARAP1|116985   16.17556 16.06011 15.56131
#&gt; ATP5C1|509     16.52061 16.49485 15.97965
#&gt; BOLA2|552900   14.96354 14.09810 14.11716
#&gt; C16orf13|84326 13.85646 13.88293 13.58011</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: 'fa0226bf1a3df7e1c8ab8f0e70450e0a',
    indexName: 'stem-cell-analysis-in-psichomics',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


