<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='Gene expression is filtered and normalised in the following steps:

Filter gene expression;
Normalise gene expression with calcNormFactors;
If performVoom = FALSE, compute counts per million (CPM) using
cpm and log2-transform values if
log2transform = TRUE;
If performVoom = TRUE, use voom to compute
log2-CPM, quantile-normalise (if method = "quantile") and estimate
mean-variance relationship to calculate observation-level weights.

'><title>Filter and normalise gene expression — normaliseGeneExpression • psichomics</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Filter and normalise gene expression — normaliseGeneExpression"><meta property="og:description" content='Gene expression is filtered and normalised in the following steps:

Filter gene expression;
Normalise gene expression with calcNormFactors;
If performVoom = FALSE, compute counts per million (CPM) using
cpm and log2-transform values if
log2transform = TRUE;
If performVoom = TRUE, use voom to compute
log2-CPM, quantile-normalise (if method = "quantile") and estimate
mean-variance relationship to calculate observation-level weights.

'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">psichomics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.20.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    <a class="dropdown-item" href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    <a class="dropdown-item" href="../articles/custom_data.html">Loading user-provided data</a>
    <a class="dropdown-item" href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nuno-agostinho/psichomics/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Filter and normalise gene expression</h1>
      <small class="dont-index">Source: <a href="https://github.com/nuno-agostinho/psichomics/blob/HEAD/R/data_geNormalisationFiltering.R" class="external-link"><code>R/data_geNormalisationFiltering.R</code></a></small>
      <div class="d-none name"><code>normaliseGeneExpression.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Gene expression is filtered and normalised in the following steps:</p>
<ul><li><p>Filter gene expression;</p></li>
<li><p>Normalise gene expression with <code><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></code>;</p></li>
<li><p>If <code>performVoom = FALSE</code>, compute counts per million (CPM) using
<code><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">cpm</a></code> and log2-transform values if
<code>log2transform = TRUE</code>;</p></li>
<li><p>If <code>performVoom = TRUE</code>, use <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom</a></code> to compute
log2-CPM, quantile-normalise (if <code>method = "quantile"</code>) and estimate
mean-variance relationship to calculate observation-level weights.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">normaliseGeneExpression</span><span class="op">(</span>
  <span class="va">geneExpr</span>,
  geneFilter <span class="op">=</span> <span class="cn">NULL</span>,
  method <span class="op">=</span> <span class="st">"TMM"</span>,
  p <span class="op">=</span> <span class="fl">0.75</span>,
  log2transform <span class="op">=</span> <span class="cn">TRUE</span>,
  priorCount <span class="op">=</span> <span class="fl">0.25</span>,
  performVoom <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>

<span class="fu">normalizeGeneExpression</span><span class="op">(</span>
  <span class="va">geneExpr</span>,
  geneFilter <span class="op">=</span> <span class="cn">NULL</span>,
  method <span class="op">=</span> <span class="st">"TMM"</span>,
  p <span class="op">=</span> <span class="fl">0.75</span>,
  log2transform <span class="op">=</span> <span class="cn">TRUE</span>,
  priorCount <span class="op">=</span> <span class="fl">0.25</span>,
  performVoom <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>geneExpr</dt>
<dd><p>Matrix or data frame: gene expression</p></dd>


<dt>geneFilter</dt>
<dd><p>Boolean: filtered genes (if <code>NULL</code>, skip filtering)</p></dd>


<dt>method</dt>
<dd><p>Character: normalisation method, including <code>TMM</code>,
<code>RLE</code>, <code>upperquartile</code>, <code>none</code> or <code>quantile</code> (see
Details)</p></dd>


<dt>p</dt>
<dd><p>numeric value between 0 and 1 specifying which quantile of the counts should be used by <code>method="upperquartile"</code>.</p></dd>


<dt>log2transform</dt>
<dd><p>Boolean: perform log2-transformation?</p></dd>


<dt>priorCount</dt>
<dd><p>Average count to add to each observation to avoid zeroes
after log-transformation</p></dd>


<dt>performVoom</dt>
<dd><p>Boolean: perform mean-variance modelling
(using <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom</a></code>)?</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Filtered and normalised gene expression</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">edgeR::calcNormFactors</a></code> will be used to normalise gene
expression if <code>method</code> is <code>TMM</code>, <code>RLE</code>, <code>upperquartile</code>
or <code>none</code>. If <code>performVoom = TRUE</code>, <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom</a></code> will
only normalise if <code>method = "quantile"</code>.</p>
<p>Available normalisation methods:</p><ul><li><p><code>TMM</code> is recommended for most RNA-seq data where more than half of
the genes are believed not differentially expressed between any pair of
samples;</p></li>
<li><p><code>RLE</code> calculates the median library from the geometric mean of all
columns and the median ratio of each sample to the median library is taken as
the scale factor;</p></li>
<li><p><code>upperquartile</code> calculates the scale factors from a given quantile
of the counts for each library, after removing genes with zero counts in all
libraries;</p></li>
<li><p><code>quantile</code> forces the entire empirical distribution of each
column to be identical (only performed if <code>performVoom = TRUE</code>).</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other functions for gene expression pre-processing: 
<code><a href="convertGeneIdentifiers.html">convertGeneIdentifiers</a>()</code>,
<code><a href="filterGeneExpr.html">filterGeneExpr</a>()</code>,
<code><a href="plotGeneExprPerSample.html">plotGeneExprPerSample</a>()</code>,
<code><a href="plotLibrarySize.html">plotLibrarySize</a>()</code>,
<code><a href="plotRowStats.html">plotRowStats</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="va">geneExpr</span> <span class="op">&lt;-</span> <span class="fu"><a href="readFile.html">readFile</a></span><span class="op">(</span><span class="st">"ex_gene_expression.RDS"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu">normaliseGeneExpression</span><span class="op">(</span><span class="va">geneExpr</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                Normal 1 Normal 2 Normal 3 Normal 4 Normal 5 Cancer 1 Cancer 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ACTN1|87       16.44383 16.78304 16.48223 17.24207 17.07503 17.10327 16.49269</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ADAM15|8751    16.70099 16.09395 16.28898 15.96189 15.31259 15.63044 15.97064</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AKAP8L|26993   14.14424 14.80911 14.24193 14.41006 15.30310 14.58999 14.11916</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AKR1A1|10327   15.14212 15.06161 14.85207 15.74210 15.51387 15.57613 15.64140</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ALDOA|226      18.58900 20.00823 18.57857 18.85194 18.54547 18.33072 18.78489</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ANXA6|309      16.80155 16.85176 16.67479 16.76024 16.67841 16.79276 16.81781</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ARAP1|116985   16.01864 15.57970 15.45777 15.44823 15.82390 15.95734 15.89010</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP5C1|509     14.74478 15.47092 16.10871 15.23351 15.62633 15.44610 15.49723</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BOLA2|552900   14.03579 15.05401 14.67957 13.53149 14.32091 14.37295 14.08933</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> C16orf13|84326 13.75582 14.21114 14.03332 13.69816 14.23023 13.78593 13.60679</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                Cancer 3 Cancer 4 Cancer 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ACTN1|87       16.81314 18.48316 17.61675</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ADAM15|8751    16.39172 15.97958 16.13043</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AKAP8L|26993   15.14118 14.98893 13.49192</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AKR1A1|10327   15.32114 15.90415 14.74053</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ALDOA|226      18.40211 18.50496 18.54164</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ANXA6|309      17.12530 16.53170 17.37841</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ARAP1|116985   16.17556 16.06011 15.56131</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP5C1|509     16.52061 16.49485 15.97965</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BOLA2|552900   14.96354 14.09810 14.11716</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> C16orf13|84326 13.85646 13.88293 13.58011</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

    </footer></div>

  

  

  </body></html>

