<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Preparing an Alternative Splicing Annotation for psichomics • psichomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Preparing an Alternative Splicing Annotation for psichomics">
<meta property="og:description" content="psichomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">psichomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.18.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    </li>
    <li>
      <a href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    </li>
    <li>
      <a href="../articles/custom_data.html">Loading user-provided data</a>
    </li>
    <li>
      <a href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nuno-agostinho/psichomics/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="AS_events_preparation_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Preparing an Alternative Splicing Annotation for psichomics</h1>
                        <h4 class="author">Nuno Saraiva Agostinho</h4>
            
            <h4 class="date">2021-10-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/nuno-agostinho/psichomics/blob/master/vignettes/AS_events_preparation.Rmd"><code>vignettes/AS_events_preparation.Rmd</code></a></small>
      <div class="hidden name"><code>AS_events_preparation.Rmd</code></div>

    </div>

    
    
<div id="creating-custom-alternative-splicing-annotation" class="section level1">
<h1 class="hasAnchor">
<a href="#creating-custom-alternative-splicing-annotation" class="anchor"></a>Creating custom alternative splicing annotation</h1>
<p>psichomics quantifies alternative splicing based on alternative splicing event annotations from <a href="http://genes.mit.edu/burgelab/miso/">MISO</a>, <a href="https://bitbucket.org/regulatorygenomicsupf/suppa">SUPPA</a>, <a href="https://github.com/vastgroup/vast-tools">VAST-TOOLS</a> and <a href="http://rnaseq-mats.sourceforge.net">rMATS</a>. New alternative splicing annotation may be prepared and used in psichomics by parsing alternative splicing events from those tools. Please contact me if you would like to see support for other tools.</p>
<p>This tutorial will guide you on how to parse alternative splicing events from different tools. To do so, start by loading the following packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://nuno-agostinho.github.io/psichomics/">psichomics</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://had.co.nz/plyr">plyr</a></span><span class="op">)</span></code></pre></div>
<div id="suppa-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#suppa-annotation" class="anchor"></a>SUPPA annotation</h2>
<p><a href="https://bitbucket.org/regulatorygenomicsupf/suppa">SUPPA</a> generates alternative splicing events based on a transcript annotation. Start by running SUPPA’s <code>generateEvents</code> script with a transcript file (GTF format) for all event types, if desired. See <a href="https://bitbucket.org/regulatorygenomicsupf/suppa">SUPPA’s page</a> for more information.</p>
<p>The resulting output will include a directory containing tab-delimited files with alternative splicing events (one file for each event type). Hand the path of this directory to the function <code><a href="../reference/parseSuppaAnnotation.html">parseSuppaAnnotation()</a></code>, prepare the annotation using <code><a href="../reference/prepareAnnotationFromEvents.html">prepareAnnotationFromEvents()</a></code> and save the output to a RDS file:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># suppaOutput &lt;- "path/to/SUPPA/output"</span>

<span class="co"># Replace `genome` for the string with the identifier before the first</span>
<span class="co"># underscore in the filenames of that directory (for instance, if one of your </span>
<span class="co"># filenames of interest is "hg19_A3.ioe", the string would be "hg19")</span>
<span class="va">suppa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parseSuppaAnnotation.html">parseSuppaAnnotation</a></span><span class="op">(</span><span class="va">suppaOutput</span>, genome<span class="op">=</span><span class="st">"hg19"</span><span class="op">)</span>
<span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareAnnotationFromEvents.html">prepareAnnotationFromEvents</a></span><span class="op">(</span><span class="va">suppa</span><span class="op">)</span>

<span class="co"># suppaFile &lt;- "suppa_hg19_annotation.RDS"</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">annot</span>, file<span class="op">=</span><span class="va">suppaFile</span><span class="op">)</span></code></pre></div>
</div>
<div id="rmats-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#rmats-annotation" class="anchor"></a>rMATS annotation</h2>
<p>Just like <a href="https://bitbucket.org/regulatorygenomicsupf/suppa">SUPPA</a>, <a href="http://rnaseq-mats.sourceforge.net">rMATS</a> also allows to generate alternative splicing events based on a transcript annotation, although two BAM or FASTQ files are required to generate alternative splicing events. Read <a href="http://rnaseq-mats.sourceforge.net">rMATS’ page</a> for more information.</p>
<p>The resulting output of <a href="http://rnaseq-mats.sourceforge.net">rMATS</a> is then handed out to the function <code><a href="../reference/parseSuppaAnnotation.html">parseMatsAnnotation()</a></code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># matsOutput &lt;- "path/to/rMATS/output"</span>
<span class="va">mats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parseSuppaAnnotation.html">parseMatsAnnotation</a></span><span class="op">(</span>
    <span class="va">matsOutput</span>,         <span class="co"># Output directory from rMATS</span>
    genome <span class="op">=</span> <span class="st">"fromGTF"</span>, <span class="co"># Identifier of the filenames</span>
    novelEvents<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>   <span class="co"># Parse novel events?</span>
<span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareAnnotationFromEvents.html">prepareAnnotationFromEvents</a></span><span class="op">(</span><span class="va">mats</span><span class="op">)</span>

<span class="co"># matsFile &lt;- "mats_hg19_annotation.RDS"</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">annot</span>, file<span class="op">=</span><span class="va">matsFile</span><span class="op">)</span></code></pre></div>
</div>
<div id="miso-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#miso-annotation" class="anchor"></a>MISO annotation</h2>
<p>Simply retrieve <a href="https://miso.readthedocs.io/en/fastmiso/annotation.html">MISO’s alternative splicing annotation</a> and give the path to the downloaded folder as input.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># misoAnnotation &lt;- "path/to/MISO/annotation"</span>
<span class="va">miso</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parseSuppaAnnotation.html">parseMisoAnnotation</a></span><span class="op">(</span><span class="va">misoAnnotation</span><span class="op">)</span>
<span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareAnnotationFromEvents.html">prepareAnnotationFromEvents</a></span><span class="op">(</span><span class="va">miso</span><span class="op">)</span>

<span class="co"># misoFile &lt;- "miso_AS_annotation_hg19.RDS"</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">annot</span>, file<span class="op">=</span><span class="va">misoFile</span><span class="op">)</span></code></pre></div>
</div>
<div id="vast-tools-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#vast-tools-annotation" class="anchor"></a>VAST-TOOLS annotation</h2>
<p>Download and extract <a href="https://github.com/vastgroup/vast-tools#vastdb-libraries">VAST-TOOLS’ alternative splicing annotation</a> and use the path to the <code>TEMPLATES</code> subfolder as the input of <code><a href="../reference/parseSuppaAnnotation.html">parseVastToolsAnnotation()</a></code>. Complex events (i.e. alternative coordinates for the exon ends) are not currently supported.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># vastAnnotation &lt;- "path/to/VASTDB/libs/TEMPLATES"</span>
<span class="va">vast</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parseSuppaAnnotation.html">parseVastToolsAnnotation</a></span><span class="op">(</span><span class="va">vastAnnotation</span>, genome<span class="op">=</span><span class="st">"Hsa"</span><span class="op">)</span>
<span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareAnnotationFromEvents.html">prepareAnnotationFromEvents</a></span><span class="op">(</span><span class="va">vast</span><span class="op">)</span>

<span class="co"># vastFile &lt;- "vast_AS_annotation_hg19.RDS"</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">annot</span>, file<span class="op">=</span><span class="va">vastFile</span><span class="op">)</span></code></pre></div>
</div>
<div id="combining-annotation-from-different-sources" class="section level2">
<h2 class="hasAnchor">
<a href="#combining-annotation-from-different-sources" class="anchor"></a>Combining annotation from different sources</h2>
<p>To combine the annotation from different sources, provide the parsed annotations of interest simultaneously to the function <code>prepareAnnotationFromEvents</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Combine the annotation from SUPPA, MISO, rMATS and VAST-TOOLS</span>
<span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareAnnotationFromEvents.html">prepareAnnotationFromEvents</a></span><span class="op">(</span><span class="va">suppa</span>, <span class="va">vast</span>, <span class="va">mats</span>, <span class="va">miso</span><span class="op">)</span>

<span class="co"># annotFile &lt;- "AS_annotation_hg19.RDS"</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">annot</span>, file<span class="op">=</span><span class="va">annotFile</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="quantifying-alternative-splicing-using-the-custom-annotation" class="section level1">
<h1 class="hasAnchor">
<a href="#quantifying-alternative-splicing-using-the-custom-annotation" class="anchor"></a>Quantifying alternative splicing using the custom annotation</h1>
<p>The created alternative splicing annotation can be used in psichomics for alternative splicing quantification. To do so, when using the GUI version of psichomics, be sure to select the <strong>Load annotation from file…</strong> option, click the button that appears below and select the recently created RDS file.</p>
<p>Otherwise, if you are using the CLI version, perform the following steps:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="va">annotFile</span><span class="op">)</span> <span class="co"># "annotFile" is the path to the annotation file</span>
<span class="va">junctionQuant</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFile.html">readFile</a></span><span class="op">(</span><span class="st">"ex_junctionQuant.RDS"</span><span class="op">)</span> <span class="co"># example set</span>

<span class="va">psi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantifySplicing.html">quantifySplicing</a></span><span class="op">(</span><span class="va">annot</span>, <span class="va">junctionQuant</span><span class="op">)</span></code></pre></div>
<pre><code>## Using 1 of 60 events (2%) whose junctions are present in junction quantification data...</code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">psi</span> <span class="co"># may have 0 rows because of the small junction quantification set</span></code></pre></div>
<pre><code>##                                                   Normal 1  Normal 2 Normal 3
## SE_1_+_23385660_23385840_23385851_23395032_KDM1A 0.7777778 0.4444444      0.6
##                                                  Cancer 1  Cancer 2  Cancer 3
## SE_1_+_23385660_23385840_23385851_23395032_KDM1A      0.4 0.4285714 0.5555556</code></pre>
</div>
<div id="feedback" class="section level1">
<h1 class="hasAnchor">
<a href="#feedback" class="anchor"></a>Feedback</h1>
<p>All feedback on the program, documentation and associated material (including this tutorial) is welcome. Please send any suggestions and comments to:</p>
<blockquote>
<p>Nuno Saraiva-Agostinho (<a href="mailto:nunoagostinho@medicina.ulisboa.pt" class="email">nunoagostinho@medicina.ulisboa.pt</a>)</p>
<p><a href="http://imm.medicina.ulisboa.pt/group/distrans/">Disease Transcriptomics Lab, Instituto de Medicina Molecular (Portugal)</a></p>
</blockquote>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'fa0226bf1a3df7e1c8ab8f0e70450e0a',
    indexName: 'stem-cell-analysis-in-psichomics',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
