<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="psichomics">
<title>Preparing an Alternative Splicing Annotation for psichomics • psichomics</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Preparing an Alternative Splicing Annotation for psichomics">
<meta property="og:description" content="psichomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">psichomics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.28.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    <a class="dropdown-item" href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    <a class="dropdown-item" href="../articles/custom_data.html">Loading user-provided data</a>
    <a class="dropdown-item" href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nuno-agostinho/psichomics/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Preparing an Alternative Splicing Annotation for psichomics</h1>
                        <h4 data-toc-skip class="author">Nuno Saraiva
Agostinho</h4>
            
            <h4 data-toc-skip class="date">2024-03-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/nuno-agostinho/psichomics/blob/HEAD/vignettes/AS_events_preparation.Rmd" class="external-link"><code>vignettes/AS_events_preparation.Rmd</code></a></small>
      <div class="d-none name"><code>AS_events_preparation.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="creating-custom-alternative-splicing-annotation">Creating custom alternative splicing annotation<a class="anchor" aria-label="anchor" href="#creating-custom-alternative-splicing-annotation"></a>
</h2>
<p>psichomics quantifies alternative splicing based on alternative
splicing event annotations from <a href="http://genes.mit.edu/burgelab/miso/" class="external-link">MISO</a>, <a href="https://bitbucket.org/regulatorygenomicsupf/suppa" class="external-link">SUPPA</a>, <a href="https://github.com/vastgroup/vast-tools" class="external-link">VAST-TOOLS</a> and <a href="http://rnaseq-mats.sourceforge.net" class="external-link">rMATS</a>. New alternative
splicing annotation may be prepared and used in psichomics by parsing
alternative splicing events from those tools. Please contact me if you
would like to see support for other tools.</p>
<p>This tutorial will guide you on how to parse alternative splicing
events from different tools. To do so, start by loading the following
packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nuno-agostinho.github.io/psichomics/">psichomics</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://had.co.nz/plyr" class="external-link">plyr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="suppa-annotation">SUPPA annotation<a class="anchor" aria-label="anchor" href="#suppa-annotation"></a>
</h3>
<p><a href="https://bitbucket.org/regulatorygenomicsupf/suppa" class="external-link">SUPPA</a>
generates alternative splicing events based on a transcript annotation.
Start by running SUPPA’s <code>generateEvents</code> script with a
transcript file (GTF format) for all event types, if desired. See <a href="https://bitbucket.org/regulatorygenomicsupf/suppa" class="external-link">SUPPA’s
page</a> for more information.</p>
<p>The resulting output will include a directory containing
tab-delimited files with alternative splicing events (one file for each
event type). Hand the path of this directory to the function
<code><a href="../reference/parseMisoAnnotation.html">parseSuppaAnnotation()</a></code>, prepare the annotation using
<code><a href="../reference/prepareAnnotationFromEvents.html">prepareAnnotationFromEvents()</a></code> and save the output to a RDS
file:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># suppaOutput &lt;- "path/to/SUPPA/output"</span></span>
<span></span>
<span><span class="co"># Replace `genome` for the string with the identifier before the first</span></span>
<span><span class="co"># underscore in the filenames of that directory (for instance, if one of your </span></span>
<span><span class="co"># filenames of interest is "hg19_A3.ioe", the string would be "hg19")</span></span>
<span><span class="va">suppa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parseMisoAnnotation.html">parseSuppaAnnotation</a></span><span class="op">(</span><span class="va">suppaOutput</span>, genome<span class="op">=</span><span class="st">"hg19"</span><span class="op">)</span></span>
<span><span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareAnnotationFromEvents.html">prepareAnnotationFromEvents</a></span><span class="op">(</span><span class="va">suppa</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># suppaFile &lt;- "suppa_hg19_annotation.RDS"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">annot</span>, file<span class="op">=</span><span class="va">suppaFile</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="rmats-annotation">rMATS annotation<a class="anchor" aria-label="anchor" href="#rmats-annotation"></a>
</h3>
<p>Just like <a href="https://bitbucket.org/regulatorygenomicsupf/suppa" class="external-link">SUPPA</a>, <a href="http://rnaseq-mats.sourceforge.net" class="external-link">rMATS</a> also allows to
generate alternative splicing events based on a transcript annotation,
although two BAM or FASTQ files are required to generate alternative
splicing events. Read <a href="http://rnaseq-mats.sourceforge.net" class="external-link">rMATS’ page</a> for more
information.</p>
<p>The resulting output of <a href="http://rnaseq-mats.sourceforge.net" class="external-link">rMATS</a> is then handed out
to the function <code><a href="../reference/parseMisoAnnotation.html">parseMatsAnnotation()</a></code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># matsOutput &lt;- "path/to/rMATS/output"</span></span>
<span><span class="va">mats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parseMisoAnnotation.html">parseMatsAnnotation</a></span><span class="op">(</span></span>
<span>    <span class="va">matsOutput</span>,         <span class="co"># Output directory from rMATS</span></span>
<span>    genome <span class="op">=</span> <span class="st">"fromGTF"</span>, <span class="co"># Identifier of the filenames</span></span>
<span>    novelEvents<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>   <span class="co"># Parse novel events?</span></span>
<span><span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareAnnotationFromEvents.html">prepareAnnotationFromEvents</a></span><span class="op">(</span><span class="va">mats</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># matsFile &lt;- "mats_hg19_annotation.RDS"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">annot</span>, file<span class="op">=</span><span class="va">matsFile</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="miso-annotation">MISO annotation<a class="anchor" aria-label="anchor" href="#miso-annotation"></a>
</h3>
<p>Simply retrieve <a href="https://miso.readthedocs.io/en/fastmiso/annotation.html" class="external-link">MISO’s
alternative splicing annotation</a> and give the path to the downloaded
folder as input.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># misoAnnotation &lt;- "path/to/MISO/annotation"</span></span>
<span><span class="va">miso</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parseMisoAnnotation.html">parseMisoAnnotation</a></span><span class="op">(</span><span class="va">misoAnnotation</span><span class="op">)</span></span>
<span><span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareAnnotationFromEvents.html">prepareAnnotationFromEvents</a></span><span class="op">(</span><span class="va">miso</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># misoFile &lt;- "miso_AS_annotation_hg19.RDS"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">annot</span>, file<span class="op">=</span><span class="va">misoFile</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="vast-tools-annotation">VAST-TOOLS annotation<a class="anchor" aria-label="anchor" href="#vast-tools-annotation"></a>
</h3>
<p>Download and extract <a href="https://github.com/vastgroup/vast-tools#vastdb-libraries" class="external-link">VAST-TOOLS’
alternative splicing annotation</a> and use the path to the
<code>TEMPLATES</code> subfolder as the input of
<code><a href="../reference/parseMisoAnnotation.html">parseVastToolsAnnotation()</a></code>. Complex events
(i.e. alternative coordinates for the exon ends) are not currently
supported.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># vastAnnotation &lt;- "path/to/VASTDB/libs/TEMPLATES"</span></span>
<span><span class="va">vast</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parseMisoAnnotation.html">parseVastToolsAnnotation</a></span><span class="op">(</span><span class="va">vastAnnotation</span>, genome<span class="op">=</span><span class="st">"Hsa"</span><span class="op">)</span></span>
<span><span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareAnnotationFromEvents.html">prepareAnnotationFromEvents</a></span><span class="op">(</span><span class="va">vast</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># vastFile &lt;- "vast_AS_annotation_hg19.RDS"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">annot</span>, file<span class="op">=</span><span class="va">vastFile</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="combining-annotation-from-different-sources">Combining annotation from different sources<a class="anchor" aria-label="anchor" href="#combining-annotation-from-different-sources"></a>
</h3>
<p>To combine the annotation from different sources, provide the parsed
annotations of interest simultaneously to the function
<code>prepareAnnotationFromEvents</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Combine the annotation from SUPPA, MISO, rMATS and VAST-TOOLS</span></span>
<span><span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareAnnotationFromEvents.html">prepareAnnotationFromEvents</a></span><span class="op">(</span><span class="va">suppa</span>, <span class="va">vast</span>, <span class="va">mats</span>, <span class="va">miso</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># annotFile &lt;- "AS_annotation_hg19.RDS"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">annot</span>, file<span class="op">=</span><span class="va">annotFile</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="quantifying-alternative-splicing-using-the-custom-annotation">Quantifying alternative splicing using the custom annotation<a class="anchor" aria-label="anchor" href="#quantifying-alternative-splicing-using-the-custom-annotation"></a>
</h2>
<p>The created alternative splicing annotation can be used in psichomics
for alternative splicing quantification. To do so, when using the GUI
version of psichomics, be sure to select the <strong>Load annotation
from file…</strong> option, click the button that appears below and
select the recently created RDS file.</p>
<p>Otherwise, if you are using the CLI version, perform the following
steps:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">annotFile</span><span class="op">)</span> <span class="co"># "annotFile" is the path to the annotation file</span></span>
<span><span class="va">junctionQuant</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFile.html">readFile</a></span><span class="op">(</span><span class="st">"ex_junctionQuant.RDS"</span><span class="op">)</span> <span class="co"># example set</span></span>
<span></span>
<span><span class="va">psi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantifySplicing.html">quantifySplicing</a></span><span class="op">(</span><span class="va">annot</span>, <span class="va">junctionQuant</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Using 1 of 60 events (2%) whose junctions are present in junction quantification data...</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psi</span> <span class="co"># may have 0 rows because of the small junction quantification set</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                   Normal 1  Normal 2 Normal 3</span></span>
<span><span class="co">## SE_1_+_23385660_23385840_23385851_23395032_KDM1A 0.7777778 0.4444444      0.6</span></span>
<span><span class="co">##                                                  Cancer 1  Cancer 2  Cancer 3</span></span>
<span><span class="co">## SE_1_+_23385660_23385840_23385851_23395032_KDM1A      0.4 0.4285714 0.5555556</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="feedback">Feedback<a class="anchor" aria-label="anchor" href="#feedback"></a>
</h2>
<p>All feedback on the program, documentation and associated material
(including this tutorial) is welcome. Please send any suggestions and
comments to:</p>
<blockquote>
<p>Nuno Saraiva-Agostinho (<a href="mailto:nunoagostinho@medicina.ulisboa.pt" class="email">nunoagostinho@medicina.ulisboa.pt</a>)</p>
<p><a href="http://imm.medicina.ulisboa.pt/group/distrans/" class="external-link">Disease
Transcriptomics Lab, Instituto de Medicina Molecular (Portugal)</a></p>
</blockquote>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
