<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="psichomics">
<title>Loading user-provided data • psichomics</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Loading user-provided data">
<meta property="og:description" content="psichomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">psichomics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.24.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    <a class="dropdown-item" href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    <a class="dropdown-item" href="../articles/custom_data.html">Loading user-provided data</a>
    <a class="dropdown-item" href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nuno-agostinho/psichomics/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="custom_data_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Loading user-provided data</h1>
                        <h4 data-toc-skip class="author">Nuno Saraiva-Agostinho</h4>
            
            <h4 data-toc-skip class="date">2022-11-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/nuno-agostinho/psichomics/blob/HEAD/vignettes/custom_data.Rmd" class="external-link"><code>vignettes/custom_data.Rmd</code></a></small>
      <div class="d-none name"><code>custom_data.Rmd</code></div>
    </div>

    
    
<p><em>psichomics</em> is an interactive R package for integrative analyses of alternative splicing and gene expression based on data from multiple sources, including user-provided data.</p>
<div class="section level2">
<h2 id="supported-file-formats">Supported file formats<a class="anchor" aria-label="anchor" href="#supported-file-formats"></a>
</h2>
<p>psichomics supports the following data sources. Each source has a link with instructions for data loading.</p>
<table style="width:100%;" class="table">
<colgroup>
<col width="57%">
<col width="7%">
<col width="7%">
<col width="7%">
<col width="7%">
<col width="14%">
</colgroup>
<thead><tr class="header">
<th align="left">Source</th>
<th align="center">Sample information</th>
<th align="center">Subject information</th>
<th align="center">Gene expression</th>
<th align="center">Exon-exon junction quantification</th>
<th align="center">Alternative splicing quantification</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><a href="#prepare-sra-run-selector-data">SRA Run Selector</a></td>
<td align="center">Yes</td>
<td align="center"></td>
<td align="center"></td>
<td align="center"></td>
<td align="center"></td>
</tr>
<tr class="even">
<td align="left"><a href="#prepare-tables-based-on-rna-seq-data-using-star">STAR</a></td>
<td align="center"></td>
<td align="center"></td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center"></td>
</tr>
<tr class="odd">
<td align="left"><a href="#prepare-vast-tools-data">VAST-TOOLS</a></td>
<td align="center"></td>
<td align="center"></td>
<td align="center">Yes</td>
<td align="center"></td>
<td align="center">Yes</td>
</tr>
<tr class="even">
<td align="left"><a href="#prepare-firebrowse-data">TCGA (via FireBrowse)</a></td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center"></td>
</tr>
<tr class="odd">
<td align="left"><a href="https://nuno-agostinho.github.io/psichomics/articles/CLI_tutorial.html#load-sra-project-data-using-recount">SRA (via recount)</a></td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center"></td>
</tr>
<tr class="even">
<td align="left"><a href="#prepare-gtex-data">GTEx</a></td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center"></td>
</tr>
<tr class="odd">
<td align="left"><a href="#prepare-data-from-any-source">Other sources</a></td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Limited*</td>
</tr>
</tbody>
</table>
<p>* psichomics cannot fully parse alternative splicing events (e.g. it may not identify the cognate gene and coordinates) based on tables from these sources.</p>
<div class="section level3">
<h3 id="prepare-sra-run-selector-data">Prepare <a href="https://www.ncbi.nlm.nih.gov/Traces/study/" class="external-link">SRA Run Selector</a> data<a class="anchor" aria-label="anchor" href="#prepare-sra-run-selector-data"></a>
</h3>
<p>The <a href="https://www.ncbi.nlm.nih.gov/Traces/study/" class="external-link">SRA Run Selector</a> contains sample metadata that can be downloaded for all or selected samples from a SRA project. To download sample information, click the <strong>Metadata</strong> button in the <strong>Download</strong> columns. The output file is usually named <code>SraRunTable.txt</code>.</p>
<p>To proceed loading the data, move the downloaded file to a new folder and follow the instructions in <a href="#load-user-provided-data-into-psichomics">Load user-provided data into psichomics</a>.</p>
</div>
<div class="section level3">
<h3 id="prepare-tables-based-on-rna-seq-data-using-star">Prepare tables based on RNA-seq data using STAR<a class="anchor" aria-label="anchor" href="#prepare-tables-based-on-rna-seq-data-using-star"></a>
</h3>
<p>The following section goes through the steps required to load data based on RNA-seq data:</p>
<ol style="list-style-type: decimal">
<li>Retrieve FASTQ files and sample-associated information (optional if you already have the FASTQ files);</li>
<li>Map RNA-seq reads from the FASTQ files against a genome of reference using a splice-aware aligner, such as <a href="https://github.com/alexdobin/STAR" class="external-link">STAR</a>;</li>
<li>Merge and prepare its output to be correctly interpreted by psichomics;</li>
<li>Load data into psichomics.</li>
</ol>
<div class="section level4">
<h4 id="download-fastq-files-optional">Download FASTQ files (optional)<a class="anchor" aria-label="anchor" href="#download-fastq-files-optional"></a>
</h4>
<p><a href="https://www.ncbi.nlm.nih.gov/sra" class="external-link">SRA</a> is a repository of biological sequences that stores data from many published articles with the potential to answer pressing biological questions.</p>
<blockquote>
<p>The latest versions of psichomics support <strong>automatic downloading of SRA data</strong> from <a href="https://jhubiostatistics.shinyapps.io/recount/" class="external-link">recount</a>, a resource of pre-processed data for thousands of SRA projects (including gene read counts, splice junction quantification and sample metadata). First, check if the project of your interest is available in recount, thus making it quicker to analyse gene expression and alternative splicing for your samples of interest.</p>
</blockquote>
<p>Data from SRA can be downloaded using the <a href="https://github.com/ncbi/sra-tools/wiki/HowTo:-fasterq-dump" class="external-link">fasterq-dump</a> command from <a href="https://github.com/ncbi/sra-tools" class="external-link">sra-tools</a>. For instance, to retrieve samples from the <a href="https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP126561" class="external-link">SRP126561</a> project:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># List SRA samples</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="va">samples=(</span>SRR6368612 SRR6368613 SRR6368614 SRR6368615 SRR6368616 SRR6368617<span class="va">)</span></span>
<span id="cb1-3"><a href="#cb1-3"></a></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="co"># Download samples</span></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="ex">fasterq-dump</span> --split-3 <span class="va">${samples}</span></span></code></pre></div>
<blockquote>
<p><strong><code>--split-3</code></strong> allows to output one or two FASTQ files for single-end or paired-end sequencing, respectively (a third FASTQ file may also be returned containing orphaned single-end reads obtained from paired-end sequencing data)</p>
</blockquote>
<p>Sample-associated data is also available from the <a href="https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP126561" class="external-link">Run Selector</a> page. Click <strong>RunInfo Table</strong> to download the whole metadata table for all samples (usually downloaded in a file named <code>SraRunTable.txt</code>).</p>
</div>
<div class="section level4">
<h4 id="align-rna-seq-data-to-quantify-splice-junctions">Align RNA-seq data to quantify splice junctions<a class="anchor" aria-label="anchor" href="#align-rna-seq-data-to-quantify-splice-junctions"></a>
</h4>
<p>The quantification of each alternative splicing event is based on the proportion of junction reads that support the inclusion isoform, known as percent spliced-in or PSI <span class="citation">(Wang <em>et al.</em>, 2008)</span>.</p>
<p>To estimate this value for each splicing event, both alternative splicing annotation and quantification of RNA-Seq reads aligning to splice junctions (junction quantification) are required. While alternative splicing annotation is provided by the package, junction quantification will need to be prepared from user-provided data by aligning the RNA-seq reads from FASTQ files to a genome of reference. As junction reads are required to quantify alternative splicing, a splice-aware aligner will be used.</p>
<p>psichomics currently supports <a href="https://github.com/alexdobin/STAR" class="external-link">STAR</a> output.</p>
<div class="section level5">
<h5 id="index-the-genome-using-star">Index the genome using STAR<a class="anchor" aria-label="anchor" href="#index-the-genome-using-star"></a>
</h5>
<p>Before aligning FASTQ samples against a genome of reference, an index needs to be prepared.</p>
<p>Start by downloading a FASTA file of the whole genome and a GTF file with annotated transcripts. This command makes use of these <a href="https://grch37.ensembl.org/info/data/ftp/" class="external-link">human FASTA and GTF files (hg19 assembly)</a>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">mkdir</span> hg19_STAR</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="ex">STAR</span> --runMode genomeGenerate \</span>
<span id="cb2-3"><a href="#cb2-3"></a>     --genomeDir hg19_STAR \</span>
<span id="cb2-4"><a href="#cb2-4"></a>     --genomeFastaFiles /path/to/hg19.fa \</span>
<span id="cb2-5"><a href="#cb2-5"></a>     --sjdbGTFfile /path/to/hg19.gtf \</span>
<span id="cb2-6"><a href="#cb2-6"></a>     --runThreadN 4</span>
<span id="cb2-7"><a href="#cb2-7"></a></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="co"># Arguments:</span></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="co">#     --runMode             Generate the genome index</span></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="co">#     --genomeDir           Path to genome index (output)</span></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="co">#     --genomeFastaFiles    Path to genome FASTA file(s)</span></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="co">#     --sjdbGTFfile         Path to junction GTF annotation</span></span>
<span id="cb2-13"><a href="#cb2-13"></a><span class="co">#     --runThreadN 4        Run in parallel using 4 threads</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="align-against-genome-index-using-star">Align against genome index using STAR<a class="anchor" aria-label="anchor" href="#align-against-genome-index-using-star"></a>
</h5>
<p>After the genome index is generated, the sequences in the FASTQ files need to be aligned against the annotated gene and splice junctions from the previously prepared reference. The following commands make STAR output both gene and junction read counts into files ending in <code>ReadsPerGene.out.tab</code> and <code>SJ.out.tab</code>, respectively.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">align ()</span> <span class="kw">{</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>    <span class="bu">echo</span> <span class="st">"Aligning </span><span class="va">${1}</span><span class="st"> using STAR..."</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>    <span class="ex">STAR</span> --readFilesIn <span class="va">${1}</span>_1.fastq <span class="va">${1}</span>_2.fastq \</span>
<span id="cb3-4"><a href="#cb3-4"></a>         --runThreadN 16 \</span>
<span id="cb3-5"><a href="#cb3-5"></a>         --genomeDir hg19_STAR \</span>
<span id="cb3-6"><a href="#cb3-6"></a>         --readFilesCommand zcat \</span>
<span id="cb3-7"><a href="#cb3-7"></a>         --quantMode GeneCounts \</span>
<span id="cb3-8"><a href="#cb3-8"></a>         --outFileNamePrefix <span class="va">${1}</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="kw">}</span></span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="co"># Arguments:</span></span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="co">#     --readFilesIn              FASTQ files to align</span></span>
<span id="cb3-12"><a href="#cb3-12"></a><span class="co">#     --runThreadN 16            Run in parallel using 16 threads</span></span>
<span id="cb3-13"><a href="#cb3-13"></a><span class="co">#     --genomeDir                Path to genome index (input)</span></span>
<span id="cb3-14"><a href="#cb3-14"></a><span class="co">#     --readFilesCommand zcat    Use zcat to extract compressed files</span></span>
<span id="cb3-15"><a href="#cb3-15"></a><span class="co">#     --quantMode                Return gene read counts</span></span>
<span id="cb3-16"><a href="#cb3-16"></a><span class="co">#     --outFileNamePrefix        Prefix for output files</span></span>
<span id="cb3-17"><a href="#cb3-17"></a>         </span>
<span id="cb3-18"><a href="#cb3-18"></a><span class="kw">for</span> <span class="ex">each</span> in <span class="va">${samples}</span><span class="kw">;</span> <span class="kw">do</span></span>
<span id="cb3-19"><a href="#cb3-19"></a>    <span class="ex">align</span> <span class="st">"</span><span class="va">${each}</span><span class="st">"</span></span>
<span id="cb3-20"><a href="#cb3-20"></a><span class="kw">done</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="prepare-output-for-psichomics">Prepare output for psichomics<a class="anchor" aria-label="anchor" href="#prepare-output-for-psichomics"></a>
</h4>
<p>To process the resulting data files, type in R:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Change working directory to where the STAR output is</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"/path/to/aligned/output/"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nuno-agostinho.github.io/psichomics/">psichomics</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/prepareSRAmetadata.html">prepareGeneQuant</a></span><span class="op">(</span></span>
<span>    <span class="st">"SRR6368612ReadsPerGene.out.tab"</span>, <span class="st">"SRR6368613ReadsPerGene.out.tab"</span>,</span>
<span>    <span class="st">"SRR6368614ReadsPerGene.out.tab"</span>, <span class="st">"SRR6368615ReadsPerGene.out.tab"</span>,</span>
<span>    <span class="st">"SRR6368616ReadsPerGene.out.tab"</span>, <span class="st">"SRR6368617ReadsPerGene.out.tab"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/prepareSRAmetadata.html">prepareJunctionQuant</a></span><span class="op">(</span><span class="st">"SRR6368612SJ.out.tab"</span>, <span class="st">"SRR6368613SJ.out.tab"</span>, </span>
<span>                     <span class="st">"SRR6368614SJ.out.tab"</span>, <span class="st">"SRR6368615SJ.out.tab"</span>,</span>
<span>                     <span class="st">"SRR6368616SJ.out.tab"</span>, <span class="st">"SRR6368617SJ.out.tab"</span><span class="op">)</span></span></code></pre></div>
<p>To load the data, move the files (including the SRA metadata) to a new folder and follow the instructions in <a href="#load-user-provided-data-into-psichomics">Load user-provided data into psichomics</a>.</p>
</div>
</div>
<div class="section level3">
<h3 id="prepare-vast-tools-data">Prepare <a href="https://github.com/vastgroup/vast-tools" class="external-link">VAST-TOOLS</a> data<a class="anchor" aria-label="anchor" href="#prepare-vast-tools-data"></a>
</h3>
<p>psichomics supports loading inclusion levels and gene expression tables from <a href="https://github.com/vastgroup/vast-tools" class="external-link">VAST-TOOLS</a> (the tables available after running <code>vast-tools combine</code>). Note:</p>
<ul>
<li>If you are interested in analysing gene expression, please run <code>vast-tools align</code> with argument <code>--expr</code>;</li>
<li>Gene expression is returned in cRPKMs; to get both cRPKMs and gene read counts, please run <code>vast-tools combine</code> with argument <code>-C</code> (in case of doubt, always calculate both cRPKMs and gene read counts).</li>
</ul>
<p>Any sample and/or subject information may also be useful to load. Unless the sample metadata comes from <a href="https://www.ncbi.nlm.nih.gov/Traces/study/" class="external-link">SRA Run Selector</a>, please ensure that the table is recognised by psichomics: read <a href="#prepare-data-from-any-source">Prepare generic data</a>.</p>
<p>To load the data and move all files to a new folder (VAST-TOOLS alternative splicing quantification and gene expression tables and sample/subject-associated information).</p>
<p>Follow the instructions in <a href="#load-user-provided-data-into-psichomics">Load user-provided data into psichomics</a> to load the files in the visual interface. Otherwise, use function <a href="https://nuno-agostinho.github.io/psichomics/reference/loadLocalFiles"><code>loadLocalFiles()</code></a> with the folder path as an argument:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nuno-agostinho.github.io/psichomics/">psichomics</a></span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadLocalFiles.html">loadLocalFiles</a></span><span class="op">(</span><span class="st">"/path/to/psichomics/input"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">junctionQuant</span>  <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">`Junction quantification`</span></span>
<span><span class="va">sampleInfo</span>     <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">`Sample metadata`</span></span>
<span><span class="co"># Both gene read counts and cRPKMs are loaded as separate data frames</span></span>
<span><span class="va">geneReadCounts</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">`Gene expression (read counts)`</span></span>
<span><span class="va">cRPKM</span>          <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">`Gene expression (cRPKM)`</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="prepare-firebrowse-data">Prepare <a href="http://firebrowse.org" class="external-link">FireBrowse</a> data<a class="anchor" aria-label="anchor" href="#prepare-firebrowse-data"></a>
</h3>
<p><a href="http://firebrowse.org" class="external-link">FireBrowse</a> contains <a href="https://tcga-data.nci.nih.gov/docs/publications/tcga" class="external-link">TCGA</a> data for multiple tumour types and can be <a href="https://nuno-agostinho.github.io/psichomics/articles/CLI_tutorial.html#downloading-and-loading-tcga-data">automatically downloaded and then loaded</a> using psichomics.</p>
<p>Alternatively, manually downloaded files from <a href="http://firebrowse.org" class="external-link">FireBrowse</a> can be moved to a folder and then loaded in psichomics by following the instructions in <a href="#load-user-provided-data-into-psichomics">Load user-provided data into psichomics</a>.</p>
</div>
<div class="section level3">
<h3 id="prepare-gtex-data">Prepare <a href="https://www.gtexportal.org" class="external-link">GTEx</a> data<a class="anchor" aria-label="anchor" href="#prepare-gtex-data"></a>
</h3>
<p><a href="https://www.gtexportal.org" class="external-link">GTEx</a> contains data for multiple normal tissues. GTEx data can be <a href="https://nuno-agostinho.github.io/psichomics/articles/CLI_tutorial.html#load-gtex-data">automatically downloaded and then loaded</a> using psichomics.</p>
<p>Alternatively, manually downloaded files from <a href="https://www.gtexportal.org" class="external-link">GTEx</a> can be moved to a folder and then loaded in psichomics by following the instructions in <a href="#load-user-provided-data-into-psichomics">Load user-provided data into psichomics</a>.</p>
</div>
<div class="section level3">
<h3 id="prepare-data-from-any-source">Prepare data from any source<a class="anchor" aria-label="anchor" href="#prepare-data-from-any-source"></a>
</h3>
<p>psichomics supports importing generic data from any source as long as the tables are prepared as detailed below.</p>
<blockquote>
<p>Please make sure that sample and subject identifiers are exactly the same between <strong>all datasets</strong>.</p>
</blockquote>
<div class="section level4">
<h4 id="sample-information">Sample information<a class="anchor" aria-label="anchor" href="#sample-information"></a>
</h4>
<blockquote>
<p>If you are working with sample metadata from <strong>SRA Run Selector</strong>, see <a href="#prepare-sra-run-selector-data">how to prepare SRA Run Selector data</a>.</p>
</blockquote>
<ul>
<li>Tab-separated values (TSV)</li>
<li>
<strong>Sample identifiers (rows)</strong> and their <strong>attributes (columns)</strong>
</li>
<li>The first column must contain sample identifiers and be named <code>Sample ID</code>
</li>
<li>Optionally, indicate the subject associated to each sample in a column named <code>Subject ID</code> (subject identifiers must be the same as the ones used in subject information)</li>
<li>Example of a valid sample information dataset:</li>
</ul>
<table class="table">
<thead><tr class="header">
<th>Sample ID</th>
<th>Type</th>
<th>Tissue</th>
<th>Subject ID</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>SMP-01</td>
<td>Tumour</td>
<td>Lung</td>
<td>SUBJ-03</td>
</tr>
<tr class="even">
<td>SMP-02</td>
<td>Normal</td>
<td>Blood</td>
<td>SUBJ-12</td>
</tr>
<tr class="odd">
<td>SMP-03</td>
<td>Normal</td>
<td>Blood</td>
<td>SUBJ-25</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="subject-information">Subject information<a class="anchor" aria-label="anchor" href="#subject-information"></a>
</h4>
<ul>
<li>Tab-separated values (TSV)</li>
<li>
<strong>Subject identifiers (rows)</strong> and their <strong>attributes (columns)</strong>
</li>
<li>The first column must contain subject identifiers and be named <code>Subject ID</code>
</li>
<li>Example of a valid subject information dataset:</li>
</ul>
<table class="table">
<thead><tr class="header">
<th>Subject ID</th>
<th align="right">Age</th>
<th>Gender</th>
<th>Race</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>SUBJ-01</td>
<td align="right">34</td>
<td>Female</td>
<td>Black</td>
</tr>
<tr class="even">
<td>SUBJ-02</td>
<td align="right">22</td>
<td>Male</td>
<td>Black</td>
</tr>
<tr class="odd">
<td>SUBJ-03</td>
<td align="right">58</td>
<td>Female</td>
<td>Asian</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="gene-expression">Gene expression<a class="anchor" aria-label="anchor" href="#gene-expression"></a>
</h4>
<ul>
<li>Tab-separated values (TSV)</li>
<li>Read counts of <strong>genes (rows)</strong> across <strong>sample (columns)</strong> (sample identifiers must be the same as the ones used in sample information)</li>
<li>The first column must contain unique gene names (symbols, Ensembl ID, etc.) and be named <code>Gene ID</code>
</li>
<li>Example of a valid gene expression dataset:</li>
</ul>
<table class="table">
<thead><tr class="header">
<th>Gene ID</th>
<th align="right">SMP-18</th>
<th align="right">SMP-03</th>
<th align="right">SMP-54</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>AMP1</td>
<td align="right">24</td>
<td align="right">10</td>
<td align="right">43</td>
</tr>
<tr class="even">
<td>BRCA1</td>
<td align="right">38</td>
<td align="right">46</td>
<td align="right">32</td>
</tr>
<tr class="odd">
<td>BRCA2</td>
<td align="right">43</td>
<td align="right">65</td>
<td align="right">21</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="exon-exon-junction-quantification">Exon-exon junction quantification<a class="anchor" aria-label="anchor" href="#exon-exon-junction-quantification"></a>
</h4>
<ul>
<li>Tab-separated values (TSV)</li>
<li>Read counts of <strong>exon-exon junctions (rows)</strong> across <strong>samples (columns)</strong> (sample identifiers must be the same as the ones used in sample information)</li>
<li>The first column must contain junction identifiers and be named <code>Junction ID</code>
</li>
<li>Only chromosome number and capital letters X, Y, Z, W, and M are supported, followed by the genomic regions; acceptable junction identifiers include:
<ul>
<li><code>10_18748_21822</code></li>
<li><code>chromosome 10 (18748 to 21822)</code></li>
<li><code>chr10:18748-21822</code></li>
</ul>
</li>
<li>Optionally, indicate the strand with <code>+</code> or <code>-</code> at the end of the junction identifier:
<ul>
<li><code>10:3213:9402:+</code></li>
<li><code>chr10:3213-9402 -</code></li>
</ul>
</li>
<li>Junction identifiers whose chromosomes are <code>alt</code>, <code>random</code> or <code>Un</code> (i.e. alternative sequences) are discarded</li>
<li>Example of a valid exon-exon junction quantification dataset:</li>
</ul>
<table class="table">
<thead><tr class="header">
<th>Junction ID</th>
<th align="right">SMP-18</th>
<th align="right">SMP-03</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>10:6752-7393</td>
<td align="right">4</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>10:18748-21822</td>
<td align="right">8</td>
<td align="right">46</td>
</tr>
<tr class="odd">
<td>10:24257-25325</td>
<td align="right">83</td>
<td align="right">65</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="alternative-splicing-quantification-also-known-as-inclusion-levels">Alternative splicing quantification (also known as inclusion levels)<a class="anchor" aria-label="anchor" href="#alternative-splicing-quantification-also-known-as-inclusion-levels"></a>
</h4>
<blockquote>
<p>Note that psichomics cannot currently parse alternative splicing events (e.g. identify the cognate gene and coordinates) from generic, user-provided tables.</p>
</blockquote>
<ul>
<li>Tab-separated values (TSV)</li>
<li>Quantification values of <strong>alternative splicing events (rows)</strong> across <strong>samples (columns)</strong> (sample identifiers must be the same as the ones used in sample information)</li>
<li>The first column must be named <code>AS Event ID</code>
</li>
<li>Values between 0 and 1 or between 0 and 100: if the latter, values are automatically scaled between 0 and 1</li>
<li>Example of a valid alternative splicing quantification dataset:</li>
</ul>
<table class="table">
<thead><tr class="header">
<th>AS Event ID</th>
<th align="right">SMP-18</th>
<th align="right">SMP-03</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>someASevent001</td>
<td align="right">0.71</td>
<td align="right">0.30</td>
</tr>
<tr class="even">
<td>anotherASevent653</td>
<td align="right">0.63</td>
<td align="right">0.37</td>
</tr>
<tr class="odd">
<td>yetAnother097</td>
<td align="right">0.38</td>
<td align="right">0.62</td>
</tr>
</tbody>
</table>
<p>To load the data, move the files to a new folder and follow the instructions in <a href="#load-user-provided-data-into-psichomics">Load user-provided data into psichomics</a>.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="load-user-provided-data-into-psichomics">Load user-provided data into psichomics<a class="anchor" aria-label="anchor" href="#load-user-provided-data-into-psichomics"></a>
</h2>
<div class="section level3">
<h3 id="load-using-the-visual-interface">Load using the visual interface<a class="anchor" aria-label="anchor" href="#load-using-the-visual-interface"></a>
</h3>
<p>Start psichomics with the following commands in an R console or RStudio:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nuno-agostinho.github.io/psichomics/">psichomics</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/psichomics.html">psichomics</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Then, click <strong>Load user files</strong>. Click the <strong>Folder input</strong> tab and select the appropriate folder. Finally, click <strong>Load files</strong> to automatically scan and load all supported files from that folder.</p>
</div>
<div class="section level3">
<h3 id="load-using-the-command-line-interface-cli">Load using the command-line interface (CLI)<a class="anchor" aria-label="anchor" href="#load-using-the-command-line-interface-cli"></a>
</h3>
<p>Use function <a href="https://nuno-agostinho.github.io/psichomics/reference/loadLocalFiles"><code>loadLocalFiles()</code></a> with the folder path as an argument:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nuno-agostinho.github.io/psichomics/">psichomics</a></span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadLocalFiles.html">loadLocalFiles</a></span><span class="op">(</span><span class="st">"/path/to/psichomics/input"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">geneExpr</span>      <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">`Gene expression`</span></span>
<span><span class="va">junctionQuant</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">`Junction quantification`</span></span>
<span><span class="va">sampleInfo</span>    <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">`Sample metadata`</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="feedback">Feedback<a class="anchor" aria-label="anchor" href="#feedback"></a>
</h2>
<p>All feedback on the program, documentation and associated material (including this tutorial) is welcome. Please send any comments and questions to:</p>
<blockquote>
<p>Nuno Saraiva-Agostinho (<a href="mailto:nunoagostinho@medicina.ulisboa.pt" class="email">nunoagostinho@medicina.ulisboa.pt</a>)</p>
<p><a href="http://imm.medicina.ulisboa.pt/group/distrans/" class="external-link">Disease Transcriptomics Lab, Instituto de Medicina Molecular (Portugal)</a></p>
</blockquote>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references">
<div id="ref-wang2008">
<p>Wang,E.T. <em>et al.</em> (2008) Alternative isoform regulation in human tissue transcriptomes. <em>Nature</em>, <strong>456</strong>, 470–476.</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Nuno Saraiva-Agostinho, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
